{% extends "base.html" %}
{% from "macros.html" import info %}
{% block title %}Change log{% endblock %}
{% block content %}
	{# Different categories:
		-Minor: style changes, minor js, minor functionality (small impact on user)
		-Major: new page or functionality
		-New predictor: predictor added
		-New dataset
	#}
	<div class="w3-container">
		{{ info("[2022-12-19] - New feature : \
			gnomADv2 displays non cancer Max MAF + gnomad v2.0.1 become v2.1.1.") }}
		{{ info("[2022-12-19] - New feature : \
			An igv.js browser is now displayed in the gene page, together with the SpliceAI-visual predictions for the canonical transcript.") }}
		{{ info("[2022-12-07] - Minor : \
			Updated SpliceAI-visual rendering for duplications and also for complex indels in strand -.") }}
		{{ info("[2022-11-16] - Minor : \
			Intervar is now queried in hg38 version, which makes it available for hg38_secure trancripts.") }}
		{{ info("[2022-10-27] - New dataset : \
			dbNSFP has been updated from v4.1 to v4.3.") }}
		{{ info("[2022-10-27] - New feature : \
			LitVar has been updated to <a href='{0}' target='_blank'>LitVar2</a> and the litvar pop-up updated.".format(urls.ncbi_litvar2)) }}
		{{ info("[2022-10-27] - New feature : \
			Transcripts that are reported by VariantValidator to poorly match to hg19 are now available for annotation in MobiDetails. Only hg38 linked resources are available. In addition, 43 genes and 3900+ transcripts have been unlocked. These transcripts are marked 'hg38 secured' as depending on the localisation of the variant on the transcript, or on the transcript itself, hg19 mapping may succeed or not. hg19 failed variants lack some annotations or links:
			<ul>
				<li>no EVS link</li>
				<li>no intervar link nor section (waiting for an update)</li>
				<li>no Regulome DB link</li>
				<li>no hg19 nomenclatures</li>
				<li>no gnomADv2</li>
				<li>no hg19 defgen export</li>
				<li>no LOVD export on classification</li>
			</ul>") }}
		{{ info("[2022-08-11] - New resource : \
			Added a link to <a href='{0}' target='_blank'>deCAF</a> (deCODE Allele frequency Browser), which displays allelic frequencies from 150,000 UK genomes.".format(urls.decaf)) }}
		{{ info("[2022-07-05] - Software and transcript database update : \
			Modification of the database schema et update of the transcript coming from VariantValidator (version vvta_2022_02).") }}
		{{ info("[2022-06-28] - New feature : \
			A full transcript mutant prediction is now available in SpliceAI-visual on demand.") }}
		{{ info("[2022-06-10] - Bug fix : \
			Disabled SpliceAi Lookup button for delins variants as it does not correclty deal with this type of variants.") }}
		{{ info("[2022-06-09] - Bug fix : \
			Intergenic nucleotides were excluded from SpliceAI-visual predictions.") }}
		{{ info("[2022-05-28] - New predictor : \
			Bedgraph tracks representing spliceai raw scores are displayed in an igv.js browser. This functionality called 'SpliceAI-visual' is based on an original idea and code from Jean-Madeleine de Sainte Agathe and has been built in collaboration with him.") }}
		{{ info("[2022-04-28] - Minor : \
			MuPit called only for missense.") }}
		{{ info("[2022-03-18] - Minor : \
			Improved management of some VV warnings.") }}
		{{ info("[2022-03-18] - Major : \
			<ul> \
				<li>New API endpoint /api/create/vcf_str: generates annotations from a pseudo-VCF nomenclature (VCF string, e.g. 1-216422237-G-A). Can be used in the search engine (hg38 default or genome version can be added in the VCF string, e.g. hg38-1-216422237-G-A). Separator can be ':', '_' or '-'. Genome version supports GRCh37, GRCh38, hg19, hg38, case insensitive.</li> \
				<li>Variants lying in these genes cannot (hopefully temporarily) be accessed: C5orf60, LYST, TET2, MICALCL, UROS, SLC26A10, TCP10L2, STAG2, CRYBG3, DNAH17, EFCAB8, TREX1, SLC22A18AS.</li>
				<li>Along with the missense radar chart, the mean normalised scores for the single predictors family (e.g. SIFT) and for the meta-predictors (e.g. REVEL) are displayed.</li>") }}
		{{ info("[2022-03-03] - New data : \
			Data from the <a href='{0}summary' target='_blank'>ClinGen Specification Registry</a> is now linked in MD. You can find specific links for the relevant genes in the gene pages (redirection modal) and in the corresponding variants pages. This may help you in using expert panels ACMG criteria during the classification process.".format(urls.clingen_gene_spec_web)) }}
		{{ info("[2022-02-14] - Major : \
			Added about 60,000 transcripts following a major update of <a href='{0}' target = '_blank'>VariantValidator</a>. In addition, the submission of a variant lying in any of the proposed transcript will not be forced on canonical transcripts anymore. However, if possible, variants annotated on non-canonical transcripts by users will also be mapped on canonical transcripts (both annotations will be available). MD canonical transcript follow in most cases <a href='https://www.ncbi.nlm.nih.gov/refseq/MANE/' target='_blank'>MANE</a> annotations. Database scheme was also updated.".format(urls.variant_validator)) }}
		{{ info("[2021-10-07] - Minor : \
			Added new links to <a href='{0}' target = '_blank'>AlphaFold/EBI protein Structure Database</a> to the gene (redirection modal) and variant pages.".format(urls.alphafold)) }}
		{{ info("[2021-08-09] - New data : \
			Episignature data for 53 <em><a href='{0}' target='_blank'>KMT2A</a></em> variants are now available. See <a href='{1}32644126/' target='_blank'>this review</a> to learn more about this concept. Thanks to Pr. Lebre who brought the idea of integrating her team's results into MD.".format(url_for('md.gene', gene_symbol='KMT2A'), urls.ncbi_pubmed)) }}
		{{ info("[2021-06-28] - New tool version : \
			<a href='{0}' target = '_blank'>SPiP</a> has been updated to v2.1.".format(urls.spip)) }}
		{{ info("[2021-06-21] - several changes around SPiP: \
			<ul>
				<li>New value for parameter caller for the /api/variant/ endpoint: Setting caller to 'clispip' will now return the full results including SPiP (takes longer), while using 'cli' will return the results without SPiP (faster). The swagger page has also bee nupdated to reflect this change.</li>
				<li>SPiP results are now cached to be computed only once and enable a faster access the second (and third....) time</li>
				<li>SPiP has been upgraded to v2. v2 is much more computer intensive than v1 (i.e. is slower). Start a <a href='https://github.com/beboche/MobiDetails/discussions' target='_blank'>Discussion</a> on github if you have any comments!</li>
			</ul>") }}
		{{ info("[2021-06-18] - New predictor : \
			SpliceAi lookup API can be requested with the non-default parameter distance of 500bp for almost all variants (except complex indels). This can be useful to analyse deep intronic variants as well as large dels and dups.") }}
		{{ info("[2021-06-15] - Minor : \
			HGNC gene symbol and ID now appear in the variant nomenclatures table.") }}
		{{ info("[2021-06-08] - Major : \
			New formatting for the pdf export, which now includes charts and pubmed references, as well as other new features.") }}
		{{ info("[2021-06-02] - Minor : \
			SPiP results are now available using the /api/variant/ API endpoint.") }}
		{{ info("[2021-05-26] - Major : \
			New API endpoint /api/variant/: displays variants details as HTML pages or json files.") }}
		{{ info("[2021-01-16] - Minor : \
			New feature : lists of favourite variants can now be modified directly from the user's variants in the profile page.") }}
		{{ info("[2021-04-09] - Major : \
			New feature : authenticatd users are now able to generate lists of variants accessible form unique and permanent URLs from their favourite list of variants. Users can create several lists, delete them, and use the URL to reference the complete list of variants, e.g. in a publication. All this is available from your profile page. \
			") }}
		{{ info("[2021-04-01] - New external link : \
			new link directed to <a href='http://bioinfo.univ-rouen.fr/HExoSplice' target='_blank'>HexoSplice</a>, a tool to predict the impact of a variant on potential splicing regulatory elements. Only for substitutions.") }}
		{{ info("[2021-01-06] - New external link : \
			When a variant maps onto a 3D structure, a new link is displayed in the left panel to access directly MuPIT") }}
		{{ info("[2021-01-04] - Minor : \
			New search engine keyword: 'last' will return all variants annotated during the last 7 days") }}
		{{ info("[2020-12-01] - New predictor : \
			<a href='https://github.com/raphaelleman/SPiP' target='_blank'>SPiP</a> is now available as a splicing predictor.") }}
		{{ info("[2020-11-23] - Minor : \
			Added a link to <a href='https://panelapp.genomicsengland.co.uk' target='_blank'>PanelApp</a> in the gene page when relevant.") }}
		{{ info("[2020-11-05] - Minor : \
			Updated CADD from v1.5 to v1.6 and added direct links to CADD lookup for substitutions.") }}
		{{ info("[2020-11-04] - Minor : \
			Added detailed Intervar output (ACMG criteria).") }}
		{{ info("[2020-10-08] - Minor : \
			<ul> \
				<li>API endpoints /api/gene/ and /api/variant/create_g now deal with HGNC IDs in addition to HGNC names.</li> \
				<li>LOVD export is based on HGNC IDs instead of HGNC Names.</li> \
				<li>Improved retrieval of HGNC aliases and previous symbols via the search engine.</li> \
			</ul>") }}
		{{ info("[2020-10-01] - Minor : \
			Added new API end point to check submitted API keys.") }}
		{{ info("[2020-09-29] - Minor : \
			Pubmed links are now sorted and presented by 1st author, year, journal (title on mouseover).") }}
		{{ info("[2020-09-24] - New predictor : \
			<a href='http://lbgi.fr/mistic' target='_blank'>Mistic</a> is now available as a missense predictor.") }}
		{{ info("[2020-09-23] - New function : \
			Variants for which an ACMG class is defined are automatically submitted to <a href='http://www.lovd.nl' target='_blank'>LOVD</a>. <a href='{}#data_sharing'>Read more</a>.".format(url_for('md.about'))) }}
		{{ info("[2020-09-07] - New function : \
			Possibility to annotate variants using dbSNP rs ids (API). \
			The search engine can now be used to annotate variants not only to retrieve existing ones, using the following formats: \
			<ul> \
				<li>NM_206933.2:c.100C>T</li> \
				<li>NC_000001.11:g.216422237G>A;USH2A (hg38 only)</li> \
				<li>rs772808534</li> \
			</ul>") }}
		{{ info("[2020-08-26] - New function : \
			Added a <a href='{}'>page</a> to upload a text file for batch variant annotation.".format(url_for('upload.file_upload'))) }}
		{{ info("[2020-08-18] - New predictors : \
			<ul> \
				<li>Scores for miRNA binding predictors miRanda, TargetScan and RNAHybrid from <a href='{0}' target='_blank'>dbMTS</a> have been added for 3'UTR variants.</li> \
				<li><a href='{1}' target='_blank'>Eigen</a> scores have been added, either from  <a href='{2}' target='_blank'>dbNSFP</a> for coding variants, or from  <a href='{0}' target='_blank'>dbMTS</a> for 3' UTR variants.</li> \
			</ul>".format(urls.dbmts, urls.eigen, urls.dbnsfp)) }}
		{{ info("[2020-08-14] - Feature update : \
			LOVD effects are now reported when available. LOVD and Intervar results now appear in pdf reports.") }}
		{{ info("[2020-08-12] - Minor: \
			added CSS loader to most pages to increase the browsing comfort.") }}
		{{ info("[2020-08-11] - API changes:  \
			3 methods passed from GET to POST - also update create_vars_batch script in <a href='https://github.com/beboche/MDUtils' target='_blank'>MDUtils</a> package.") }}
		{{ info("[2020-08-11] - Minor : \
			improvements to JS/CSS to make MD more responsive (for small screens)") }}
		{{ info("[2020-07-09] - Minor : \
			<ul> \
				<li>Adapted MaxEntScan for indels > 9bp</li> \
				<li>Added link to RegulomeDB when possible</li> \
				<li>New reference/tool version table in about page and pdf export</li>\
			</ul>") }}
		{{ info("[2020-07-06] - New admin function : \
			added an 'Forgot my password function' and bug fix with MaxEntScan which could lead to a 500 error.") }}
		{{ info("[2020-06-24] - New API functions : \
			ablility to create variants from genomic coordinates and a HGNC gene name. Ability to update a variant's ACMG class.") }}
		{{ info("[2020-06-20] - <a href='https://cftr.iurc.montp.inserm.fr/cftr/' target='_blank'>CFTR-France</a> and MobiDetails are now working together: All variants in CFTR-France have been integrated into MobiDetails!") }}
		{{ info("[2020-05-25] - New predictor : \
			MaxEntScan is now available as a splicing predictor.") }}
		{{ info("[2020-05-13] - Minor : \
			LOVD instances are now identified by their names when available.") }}
		{{ info("[2020-05-07] - Minor : \
			Registering now requires an email activation.") }}
		{{ info("[2020-04-23] - New predictor : \
			<ul><li>Added <a href='https://sites.google.com/site/clinpred/' target='_blank'>ClinPred</a> as a new missense predictor (<a href='https://www.ncbi.nlm.nih.gov/pubmed/?term=30220433' target='_blank'>paper</a>).</li><li>Some minor style changes</li></ul>") }}
	</div>
	<br /><br /><br />
{% endblock %}
