{% from 'macros.html' import danger, info, link, span_color, section_title, modal_js, fa %}
{% extends "base.html" %}
{% block more_head %}
	<link rel="stylesheet" href="{{ url_for('static', filename='styles/datatables.1.11.5.min.css') }}">
	<!-- <link rel="stylesheet" href="//cdn.datatables.net/1.11.5/css/jquery.dataTables.min.css"> -->
	<script src="{{ url_for('static', filename='javascript/datatables.1.11.5.nopdfmake.min.js') }}" type="text/javascript"></script>
	<script src="{{ url_for('static', filename='javascript/pdfmake/0.2.5/pdfmake.min.js') }}" type="text/javascript"></script>
	<script src="{{ url_for('static', filename='javascript/pdfmake/0.2.5/vfs_fonts.min.js') }}" type="text/javascript"></script>
	<script src="{{ url_for('static', filename='javascript/Chart.min.js') }}" type="text/javascript"></script>
	<script src="{{ url_for('static', filename='javascript/igv.min.js') }}" type="text/javascript"></script>
	<script src="{{ url_for('static', filename='javascript/md_variant.js') }}??20221116" type="text/javascript"></script>
{% endblock %}
{% block content %}
	<!--left menu (hidden at first)-->
	<div class="w3-sidebar w3-bar-block w3-card w3-animate-left" id="smart_menu" style="max-height:86%;height:auto;">
		<span class="w3-bar-item w3-button w3-large w3-border-bottom w3-hover-blue" onclick="w3_close()" title="Close sidebar">X</span>
		{# hg19 #}
		{# deal with position for genome browsers #}
		{% if external_data.VCF.hg19.ncbiChr is not none %}
			{# we have hg19 data #}
			{% set positions = namespace(min=external_data.VCF.hg19.pos-25, max=external_data.VCF.hg19.pos+25, end=external_data.VCF.hg19.pos) %}
			{% if external_data.positions.size > 1 %}
			{% set positions.end = external_data.VCF.hg19.pos + external_data.positions.size %}
			{% set positions.max = external_data.VCF.hg19.pos + external_data.positions.size +25 %}
			{% endif %}
			<button class="w3-button w3-block w3-left-align w3-large w3-hover-blue w3-border-bottom w3-blue" onclick="myAccFunc('hg19_acc', 'hg19_icon');">hg19 <i class="fa fa-caret-down" id="hg19_icon"></i></button>
			<div class="w3-show" id="hg19_acc">
				{{ link("%schr%s:%s-%s&highlight=hg19.chr%s:%s-%s&hgsid=%s"|format(urls.ucsc_hg19, external_data.VCF.chr, positions.min, positions.max, external_data.VCF.chr, external_data.VCF.hg19.pos, positions.end, urls.ucsc_hg19_session), "hg19 UCSC")}}
				{# {{ link("%shg37&chromosome=%s&pos=%s"|format(urls.map2pdb, vari.chr, vari.pos), "hg19 Map2PDB") }} #}
				{{ link("%s&chromosome=%s&chromoStart=%s&chromoEnd=%s&x=0&y=0"|format(urls.evs, external_data.VCF.chr, external_data.VCF.hg19.pos, positions.end), "ESP6500") }}
				{# {{ link("%schr=%s&from=%s&to=%s&gts=rs%s&mk=%s:%s|rs%s"|format(urls.ncbi_1000g, external_data.VCF.chr, positions.min, positions.max, internal_data.frequenciesDatabases.dbSNPrsid, external_data.VCF.hg19.pos, external_data.VCF.hg19.pos, internal_data.frequenciesDatabases.dbSNPrsid), "1000 genomes")}} #}
				{{ link("%svariant/%s-%s-%s-%s"|format(urls.gnomad, external_data.VCF.chr, external_data.VCF.hg19.pos, external_data.VCF.hg19.ref, external_data.VCF.hg19.alt), "gnomAD v2")}}
				{% if external_data.positions.DNAType == 'substitution' %}
					{{ link("%ssnv/GRCh37-%s/%s:%s_%s_%s"|format(urls.cadd, external_tools.CADD.version, external_data.VCF.chr, external_data.VCF.hg19.pos, external_data.VCF.hg19.ref, external_data.VCF.hg19.alt), "CADD")}}
				{# {% endif %}
				{% if external_data.positions.DNAType == 'substitution' and external_data.positions.segmentStartType == 'exon' %} #}
					{% if external_data.positions.segmentStartType == 'exon' and internal_data.nomenclatures.pName != '?' and internal_data.nomenclatures.pName != 'Met1?' %}
					{{ link("%shg19_updated.v.201904&chr=%s&pos=%s&ref=%s&alt=%s"|format(urls.intervar, external_data.VCF.chr, external_data.VCF.hg19.pos, external_data.VCF.hg19.ref, external_data.VCF.hg19.alt), "hg19 InterVar")}}
					{% endif %}
				{% endif %}
				{% if external_data.positions.proteinTypeSoAccession is not none %}
				{{ link("%s?term=(((%s[Base Position for Assembly GRCh37]) AND %s[Chromosome] AND \"so %s\"[Molecular consequence])) OR ((%s):c.%s [Variant name])"|format(urls.ncbi_clinvar, external_data.VCF.hg19.pos, external_data.VCF.chr, external_data.positions.proteinTypeSoAccession, external_data.gene.symbol, internal_data.nomenclatures.cName), "Clinvar search")}}
				{% else %}
				{{ link("%s?term=(((%s[Base Position for Assembly GRCh37]) AND %s[Chromosome])) OR ((%s):c.%s [Variant name])"|format(urls.ncbi_clinvar, external_data.VCF.hg19.pos, external_data.VCF.chr, external_data.gene.symbol, internal_data.nomenclatures.cName), "Clinvar search")}}
				{% endif %}
				{% if internal_data.frequenciesDatabases.dbSNPrsid is not none %}
				{{ link("%srs%s"|format(urls.ncbi_dbsnp, internal_data.frequenciesDatabases.dbSNPrsid), "dbSNP")}}
				{{ link("%srs%s"|format(urls.regulomedb_hg19_search, internal_data.frequenciesDatabases.dbSNPrsid), "RegulomeDB") }}
				{% endif %}
			</div>
		{% endif %}
		{# hg38 #}
		{# deal with position for genome browsers #}
		{% set positions = namespace(min=external_data.VCF.hg38.pos-25, max=external_data.VCF.hg38.pos+25, end=external_data.VCF.hg38.pos) %}
		{% if external_data.positions.size > 1 %}
		{% set positions.end = external_data.VCF.hg38.pos + external_data.positions.size %}
		{% set positions.max = external_data.VCF.hg38.pos + external_data.positions.size +25 %}
		{% endif %}
		<button class="w3-button w3-block w3-left-align w3-large w3-hover-blue w3-border-bottom w3-blue" onclick="myAccFunc('hg38_acc', 'hg38_icon');">hg38 <i class="fa fa-caret-down" id="hg38_icon"></i></button>
		<div class="w3-show" id="hg38_acc">
			<!-- removed MuPit 20220511 -->
			{# <span style="display:none;" id="mupit_url">{{ urls.mupit }}</span>
			<span style="display:none;" id="mupit_link">{{ link("%s?gm=chr%s:%s"|format(urls.mupit, external_data.VCF.chr, external_data.VCF.hg38.pos), "MuPIT") }}</span> #}
			{{ link("%schr%s:%s-%s&highlight=hg38.chr%s:%s-%s&hgsid=%s"|format(urls.ucsc_hg38, external_data.VCF.chr, positions.min, positions.max, external_data.VCF.chr, external_data.VCF.hg38.pos, positions.end, urls.ucsc_hg38_session), "hg38 UCSC") }}
			{# {{ link("%shg38&chromosome=%s&pos=%s"|format(urls.map2pdb, vari.chr, vari.pos), "hg38 Map2PDB") }} #}
			{{ link("%svariant/%s-%s-%s-%s?dataset=gnomad_r3"|format(urls.gnomad, external_data.VCF.chr, external_data.VCF.hg38.pos, external_data.VCF.hg38.ref, external_data.VCF.hg38.alt), "gnomAD v3")}}
			{{ link("%s/region/chr%s:%s-%s"|format(urls.decaf, external_data.VCF.chr, external_data.VCF.hg38.pos, external_data.VCF.hg38.pos), "deCAF")}}
			{% if external_data.positions.DNAType == 'substitution' %}
			{{ link("%ssnv/GRCh38-%s/%s:%s_%s_%s"|format(urls.cadd, external_tools.CADD.version, external_data.VCF.chr, external_data.VCF.hg38.pos, external_data.VCF.hg38.ref, external_data.VCF.hg38.alt), "CADD")}}
			{{ link("%shg38&chr=%s&pos=%s&ref=%s&alt=%s"|format(urls.intervar, external_data.VCF.chr, external_data.VCF.hg38.pos, external_data.VCF.hg38.ref, external_data.VCF.hg38.alt), "hg38 InterVar")}}
			{% if external_data.positions.segmentStartType == 'exon' and internal_data.hexosplice.exonFirstNtCdnaPosition > 1 %}
			{{ link("%sresults.php?refSeq=>%s(%s)-%s%s\n%s&variation=c.%s&c=%s"|format(urls.hexosplice, external_data.gene.RefSeqTranscript, external_data.gene.symbol, external_data.positions.segmentStartType|capitalize, external_data.positions.segmentStartNumber, internal_data.hexosplice.exonSequence, internal_data.nomenclatures.cName, internal_data.hexosplice.exonFirstNtCdnaPosition), "HexoSplice") }}
			{% else %}
			{{ link("%sresults.php?refSeq=>%s(%s)-%s%s(51bp)\n%s&variation=26:%s"|format(urls.hexosplice, external_data.gene.RefSeqTranscript, external_data.gene.symbol, external_data.positions.segmentStartType|capitalize, external_data.positions.segmentStartNumber, external_data.sequences.wildType|replace(' ', ''), internal_data.positions.substitutionNature), "HexoSplice") }}
			{% endif %}
			{% endif %}
		</div>
		{% if external_data.gene.proteinSize and external_data.gene.proteinSize < 2700 %}
		{{ link("%sentry/%s"|format(urls.alphafold, external_data.gene.uniprotId), 'AlphaFold') }}
		{% elif not external_data.gene.proteinSize %}
		{{ link("%sentry/%s"|format(urls.alphafold, external_data.gene.uniprotId), 'AlphaFold') }}
		{% endif %}
		{% if external_data.gene.clingenCriteriaSpec %}
		{{ link("%sdoc/%s"|format(urls.clingen_gene_spec_web, external_data.gene.clingenCriteriaSpec), 'ClinGen') }}
		{% endif %}
		{{ link(url_for('md.gene', gene_symbol=external_data.gene.symbol), "MD %s"|format(external_data.gene.symbol)) }}
		{% if g.user %}
			{% if favourite != True %}
			<span class="w3-bar-item w3-button w3-large w3-hover-blue w3-border-bottom w3-center" id="favour_span" name='mark' onclick="favourite('{{ external_data.variantId }}', '{{ url_for('ajax.favourite') }}', '{{ csrf_token() }}');" title="Mark as favourite"><i id="favour" class="fa fa-star"></i>
</span>
			{% else %}
			<span class="w3-bar-item w3-button w3-large w3-hover-blue w3-border-bottom w3-center" id="favour_span" name='unmark' onclick="favourite('{{ external_data.variantId }}', '{{ url_for('ajax.favourite') }}', '{{ csrf_token() }}');" title="Unmark as favourite"><i id="favour" class="fa fa-star-o"></i>
</span>
			{% endif %}
		{% endif %}
	</div>
	<div class="w3-bar w3-light-grey" style="width:100%;z-index:1300;position:fixed;top:50px;" id="page_menu">
		<a href="#nomenclatures" class="w3-bar-item w3-button w3-mobile">Nomenclatures</a>
		<a href="#positions" class="w3-bar-item w3-button w3-mobile">Positions</a>
		<a href="#frequencies" class="w3-bar-item w3-button w3-mobile">Frequencies</a>
		<a href="#predictions" class="w3-bar-item w3-button w3-mobile">Predictions</a>
		{# {% if external_data.positions.DNAType == 'substitution' or external_data.splicingPredictions.spliceai_DS_AG is not none %} #}
		{# {% if external_data.positions.size < 25 %} #}
			<a href="#splicing" class="w3-bar-item w3-button w3-mobile">Splicing</a>
		{# {% endif %} #}
		{% if external_data.positions.proteinType == 'missense' %}
		<a href="#missense" class="w3-bar-item w3-button w3-mobile">Missense</a>
		{% endif %}
		{% if external_data.positions.DNAType == 'substitution' and external_data.miRNATargetSitesPredictions.mirandaCategory is not none %}
		<a href="#dbmts" class="w3-bar-item w3-button w3-mobile">miRNA target sites</a>
		{% endif %}
		{% if internal_data.episignature.ExpertACMGClassification is not none %}
		<a href="#episignature" class="w3-bar-item w3-button w3-mobile">EpiSignature</a>
		{% endif %}
		<a href="#class" class="w3-bar-item w3-button w3-mobile">Classification</a>
		<a href="#admin" class="w3-bar-item w3-button w3-mobile">Admin info</a>
	</div>
	<div id="global_content" style="margin-left:15%">
		<hr />
		<input type="hidden" name="variant_id" id="variant_id" value="{{ external_data.variantId }}" />
		<!--alternative mappings-->
		{% if internal_data.admin.otherIds is not none %}
		<div class="w3-margin w3-panel w3-pale-green w3-leftbar">
			<p>This variant has been annotated on at least one other transcript. You can check the following links:</p>
			<ul>
				{% for id in internal_data.admin.otherIds %}
				<li>
				{% autoescape false %}
					<a href="{{ url_for('api.variant', variant_id=id.feature_id, caller='browser') }}" onclick="show_loader();">{{ "%s(<em>%s</em>):c.%s" | format(id.refseq, id.gene_symbol, id.c_name) }}</a>
					{% if id.canonical %}
					<div class="w3-tag w3-round w3-green" style="padding:3px">
						<div class="w3-tag w3-round w3-green w3-border w3-border-white">
							MD canonical
						</div>
					</div>
					{% endif %}
				{% endautoescape %}
				</li>
				{% endfor %}
			</ul>
		</div>
		{% endif %}
		{% if internal_data.admin.mappedCanonical is not none %}
		{{ danger('This variant may be annotated on the current canonicam isoform (%s) using the following expression, <br /> directly in the search engine at the bottom of the page or by using the blue button below: <br />
		%s:g.%s;%s' | format(internal_data.admin.mappedCanonical, external_data.VCF.hg38.ncbiChr, internal_data.VCF.hg38.gName, external_data.gene.symbol))}}
		<div class="w3-margin w3-panel w3-pale-red w3-leftbar">
			<p>
				<span id="annnotation_launcher" class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="submit_create_var_g('{{ url_for('api.api_variant_g_create') }}', '{{ internal_data.admin.apiKey }}', '{{ external_data.variantId }}', '{{ csrf_token() }}');$('#annnotation_launcher').hide();$('#result_display').show();">Try annotation on the canonical {{ internal_data.admin.mappedCanonical }}</span>
				<span id="result_display" style="display:none;">Annotating the variant on {{  internal_data.admin.mappedCanonical }} <i class="fa fa-spinner w3-spin" style="font-size:48px"></i></span>
			</p>
		</div>
		{% endif %}
		<!--first section-->
		<div id="mobile_var_name" class="w3-center w3-blue" style="display:none;">{{ "%s:%s:c.%s"|format(external_data.gene.symbol, external_data.gene.RefSeqTranscript, internal_data.nomenclatures.cName) }}</div>
		<hr />
		{{ section_title('nomenclatures', 'Nomenclatures') }}
		{% if external_data.VCF.hg19.ncbiChr is none %}
			{{ danger('This variant was not correctly mapped onto the hg19 version of the human genome. Therefore, some links/resources are not reachable.') }}
		{% endif %}
		<div class="w3-container w3-responsive">
			<table class="w3-table w3-centered display compact" style="width:90%" id="nomenclature_table">
				<thead>
					<tr>
						<th class="w3-left-align">Features</th>
						<th class="w3-left-align" width="40%">Values</th>
						<th class="w3-left-align">Descriptions</th>
					</tr>
				</thead>
				<tbody>
					<!--HGNC gene symbol-->
					<tr>
						<td class="w3-left-align">HGNC gene symbol (ID):</td>
						<td class="w3-left-align" id="hgnc_symbol"><a href="{{ urls.hgnc_gene }}{{ external_data.gene.hgncId }}" target="_blank" title="open HGNC page for {{ external_data.gene.symbol }}"><span id="gene_symbol"><em>{{ external_data.gene.symbol }}</em></span><span>&nbsp;({{ external_data.gene.hgncId }})</span></td>
						<td class="w3-left-align"><em class="w3-small">HGNC gene symbol and corresponding ID</em></td>
					</tr>
					<!--HGVS DNA on transcript-->
					<tr>
						<td class="w3-left-align">HGVS DNA on transcript:</td>
						<td class="w3-left-align" id="nm_var">
							<a href="{{ urls.ncbi_nuccore }}{{ external_data.gene.RefSeqTranscript }}" target="_blank" title="open NCBI page for {{ external_data.gene.RefSeqTranscript }}"><span id="nm_acc">{{ external_data.gene.RefSeqTranscript }}</span></a>:<span id="c_name">c.{{ internal_data.nomenclatures.cName }}</span>
							{% if external_data.gene.canonical %}
							<div class="w3-tag w3-round w3-green" style="padding:3px">
								<div class="w3-tag w3-round w3-green w3-border w3-border-white">
									MD canonical
								</div>
							</div>
							{% endif %}
						</td>
						<td class="w3-left-align"><em class="w3-small">HGVS full nomenclature at DNA level on transcript</em></td>
					</tr>
					<!--IVS name-->
					{% if external_data.positions.segmentStartType == 'intron' %}
					<tr>
						<td class="w3-left-align">IVS name:</td>
						<td class="w3-left-align">{{ external_data.nomenclatures.ivsName }}</td>
						<td class="w3-left-align"><em class="w3-small">IVS nomenclature</em></td>
					</tr>
					{% endif %}
					<!--HGVS Protein-->
					<tr>
						<td class="w3-left-align">HGVS Protein:</td>
						<td class="w3-left-align"><a href="{{ urls.ncbi_prot }}{{ external_data.gene.RefSeqNp }}" target="_blank" title="open NCBI page for {{ external_data.gene.RefSeqNp }}">{{ external_data.gene.RefSeqNp }}</a>:<span id="hgvs_p_name">p.({{ internal_data.nomenclatures.pName }})</span></td>
						<td class="w3-left-align"><em class="w3-small">HGVS full nomenclature at protein level</em></td>
					</tr>
					<!--HGVS genomic (gene) removed-->
					{#
					{% if external_data.gene.RefSeqNg != 'NG_000000.0' and internal_data.nomenclatures.ngName is not none %}
						<tr>
							<td class="w3-left-align">HGVS genomic (gene):</td>
							<td class="w3-left-align"><a href="{{ urls.ncbi_nuccore }}{{ external_data.gene.RefSeqNg }}" target="_blank" title="open NCBI page for {{ external_data.gene.RefSeqNg }}">{{ external_data.gene.RefSeqNg }}</a>:g.{{ internal_data.nomenclatures.ngName }}</td>
							<td class="w3-left-align"><em class="w3-small">HGVS full nomenclature at genomic level (gene centered)</em></td>
						</tr>
					{% endif %}
					#}
					{% if external_data.VCF.hg19.ncbiChr is not none %}
						{# we have hg19 data #}
						<!--HGVS genomic (hg19)-->
						<tr>
							<td class="w3-left-align">HGVS genomic (hg19):</td>
							<td class="w3-left-align">chr{{ external_data.VCF.chr }}:g.<span id="hg19_g_name">{{ internal_data.VCF.hg19.gName }}</span></td>
							<td class="w3-left-align"><em class="w3-small">HGVS full nomenclature at genomic level (hg19)</em></td>
						</tr>
						<!--HGVS genomic strict (hg19)-->
						<tr>
							<td class="w3-left-align">HGVS strict genomic (hg19):</td>
							<td class="w3-left-align">{{ external_data.VCF.hg19.ncbiChr }}:g.{{ internal_data.VCF.hg19.gName}}</td>
							<td class="w3-left-align"><em class="w3-small">HGVS full strict nomenclature at genomic level (hg19)</em></td>
						</tr>
						<!--pseudo VCF (hg19)-->
						<tr>
							<td class="w3-left-align">pseudo VCF (<span id="genome_19">hg19</span>):</td>
							<td class="w3-left-align"><a href="{{ urls.variant_validator }}{{ external_data.VCF.chr }}-{{ external_data.VCF.hg19.pos }}-{{ external_data.VCF.hg19.ref }}-{{ external_data.VCF.hg19.alt }}&primary_assembly=GRCh37&alignment=splign" target="_blank" title="open VariantValidator"><span id="chrom_19">{{ external_data.VCF.chr }}</span>-<span id="pos_19">{{ external_data.VCF.hg19.pos }}</span>-<span id="ref_19">{{ external_data.VCF.hg19.ref }}</span>-<span id="alt_19">{{ external_data.VCF.hg19.alt }}</span></a></td>
							<td class="w3-left-align"><em class="w3-small">chr-pos-ref-alt (hg19)</em></td>
						</tr>
					{% endif %}
					<!--HGVS genomic (hg38)-->
					<tr>
						<td class="w3-left-align">HGVS genomic (hg38):</td>
						<td class="w3-left-align">chr{{ external_data.VCF.chr }}:g.<span id="hg38_g_name">{{ internal_data.VCF.hg38.gName }}</span></td>
						<td class="w3-left-align"><em class="w3-small">HGVS full nomenclature at genomic level (hg38)</em></td>
					</tr>
					<!--HGVS genomic strict (hg38)-->
					<tr>
						<td class="w3-left-align">HGVS strict genomic (hg38):</td>
						<td class="w3-left-align" id="hgvs_strict_genomic_hg38">{{ external_data.VCF.hg38.ncbiChr }}:g.{{ internal_data.VCF.hg38.gName }}</td>
						<td class="w3-left-align"><em class="w3-small">HGVS full strict nomenclature at genomic level (hg38)</em></td>
					</tr>
					<!--pseudo VCF (hg38)-->
					<tr>
						<td class="w3-left-align">pseudo VCF (<span id="genome_38">hg38</span>):</td>
						<td class="w3-left-align"><a href="{{ urls.variant_validator }}{{ external_data.VCF.chr }}-{{ external_data.VCF.hg38.pos }}-{{ external_data.VCF.hg38.ref }}-{{ external_data.VCF.hg38.alt }}&primary_assembly=GRCh38&alignment=splign" target="_blank" title="open VariantValidator"><span id="chrom_38">{{ external_data.VCF.chr }}</span>-<span id="pos_38">{{ external_data.VCF.hg38.pos }}</span>-<span id="ref_38">{{ external_data.VCF.hg38.ref }}</span>-<span id="alt_38">{{ external_data.VCF.hg38.alt }}</span></a></td>
						<td class="w3-left-align"><em class="w3-small">chr-pos-ref-alt (hg38)</em></td>
					</tr>
				</tbody>
			</table><br/><br/>
			<!--External links-->
			<div class="w3-row-padding">
				<!--Mutalyzer-->
				<div class="w3-col m4">
					<!-- <span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="window.open('{{ urls.mutalyzer_position_converter }}GRCh38&description=chr{{ external_data.VCF.chr }}:g.{{ internal_data.VCF.hg38.gName }}', '_blank');" title="Mutalyzer other possible mappings for the variant" style="width:100%">Mutalyzer alt mappings</span> -->
					<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="window.open('{{ urls.mutalyzer_normalizer }}{{ external_data.VCF.hg38.ncbiChr }}:g.{{ internal_data.VCF.hg38.gName }}', '_blank');" title="Mutalyzer Normalizer" style="width:100%">Mutalyzer</span>
				</div>
				<!--VariantValidator-->
				<div class="w3-col m4">
					<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="window.open('{{ urls.variant_validator }}{{ external_data.gene.RefSeqTranscript }}:c.' + encodeURIComponent('{{ internal_data.nomenclatures.cName }}') + '&primary_assembly=GRCh38&alignment=splign', '_blank');" title="VariantValidator for {{ external_data.gene.RefSeqTranscript }}:c.{{ internal_data.nomenclatures.cName }} (hg38)" style="width:100%">VariantValidator</span>
				</div>
				<!--HGVS-->
				<div class="w3-col m4">
					<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="window.open('{{ urls.hgvs }}', '_blank')" title="HGVS sequence variant nomenclature web site" style="width:100%">HGVS website</span>
				</div>
			</div>
		</div>
		<!--second section-->
		<hr />
		{{ section_title('positions', 'Positions') }}
		<div class="w3-responsive w3-container">
			<!--Position in transcript/protein-->
			<table class="w3-table w3-centered display compact" style="width:90%" id="position_table">
				<thead>
					<tr>
						<th class="w3-left-align">Features</th>
						<th class="w3-left-align" width="40%">Values</th>
						<th class="w3-left-align">Descriptions</th>
					</tr>
				</thead>
				<tbody>
					<tr><!--Position in transcript-->
						<td class="w3-left-align">Position in transcript:</td>
						{% if external_data.positions.segmentStartType == external_data.positions.segmentEndType and external_data.positions.segmentStartNumber == external_data.positions.segmentEndNumber %}
							<td class="w3-left-align">{{ external_data.positions.segmentStartType|capitalize }} {{ external_data.positions.segmentStartNumber }}</td>
						{% else %}
							<td class="w3-left-align">{{ external_data.positions.segmentStartType|capitalize }} {{ external_data.positions.segmentStartNumber }} - {{ external_data.positions.segmentEndType|capitalize }} {{ external_data.positions.segmentEndNumber }}</td>
						{% endif %}
						<td class="w3-left-align"><em class="w3-small">Exon/intron position in {{ external_data.gene.RefSeqTranscript }}</em></td>
					</tr><!--position in exon  and internal_data.nomenclatures.cName|match('^[^\*-]')-->
					{% if external_data.positions.segmentStartType == 'exon' and external_data.positions.segmentStartType == external_data.positions.segmentEndType and external_data.positions.segmentStartNumber == external_data.positions.segmentEndNumber %}
						{% if pos_splice is not none %}
							<tr>
								<td class="w3-left-align">Position / splice site</td>
								<td class="w3-left-align">{{ external_data.positions.nearestSsDist }} bp from {{ external_data.positions.nearestSsType }}</td>
								<td class="w3-left-align"><em class="w3-small">Position relative to the nearest splice site</em></td>
							</tr>
						{% endif %}
						{% if external_data.positions.proteinType != 'unknown' %}
						{{ external_data.positions.proteinSize }}
							{% if external_data.positions.aaPositionStart|int <= external_data.gene.proteinSize|int %}
								<!-- in case prot size is wrong-->
								<tr><!--protein-->
									<td class="w3-left-align">Position / protein</td>
									<td class="w3-left-align">
										<span>{{ external_data.positions.aaPositionStart }}</span>
										{% if external_data.positions.aaPositionStart != external_data.positions.aaPositionEnd %}
											<span> - {{ external_data.positions.aaPositionEnd }}</span>
										{%endif %}
									 <span> / {{ external_data.gene.proteinSize }}</span>
									</td>
									<td class="w3-left-align"><em class="w3-small">Position relative to the protein</em>
								</tr><!--protein domain-->
							{% endif %}
							<tr>
								<td class="w3-left-align">Position / domain</td>
								{% if external_data.positions.proteinDomain[0] is defined %}
									<td class="w3-left-align">
									{% for dom in external_data.positions.proteinDomain  %}
										{{ dom[0] }} ({{ dom[1] }} - {{ dom[2] }})&nbsp;&nbsp;
									{% endfor %}
									</td>
								{% else %}
									<td class="w3-left-align">Not in a UNIPROT defined domain</td>
								{% endif %}
								<td class="w3-left-align"><em class="w3-small">Position in a protein domain (UNIPROT: <a href="{{ urls.uniprot_id }}{{ external_data.gene.uniprotId }}" target="_blank" title="Visit UNIPROT {{ external_data.gene.uniprotId }}">{{ external_data.gene.uniprotId }})</a></em></td>
							</tr>
						{% endif %}
						{% if external_data.positions.metaDomednds is not none %}
							<td class="w3-left-align">Position tolerance</td>
							<td class="w3-left-align">{{ external_data.positions.metaDomednds }} : {{ span_color(external_data.positions.metaDomeTolerance, internal_data.positions.metaDomeColor) }} - <a href = "https://stuart.radboudumc.nl/metadome/dashboard/transcript/{{ external_data.positions.metadomeTranscript }}/" target="_blank">Transcript view</a></td>
							<td class="w3-left-align"><em class="w3-small"><a href="{{ urls.metadome }}" target="_blank" title="Visit Metadome">MetaDome</a> score, the closer to 0, the more intolerant to variation</em></td>
						{% endif %}
					{% endif %}
					<tr>
						<td class="w3-left-align">Wild type sequence</td>
						<td class="w3-left-align txt"><textarea id="wt_seq" rows="3" cols="30" readonly>{{ external_data.sequences.wildType }}</textarea></td>
						<td class="w3-left-align"><em class="w3-small">Wild type DNA sequence +/- 25 bp&nbsp;&nbsp;&nbsp;<button onclick="copy_text('wt_seq');" title="copy wild type sequence to clipboard"><i class="fa fa-clipboard" aria-hidden="true"></i></button></em></td>
					</tr>
					<tr>
						<td class="w3-left-align">Mutant sequence</td>
						<td class="w3-left-align txt"><textarea id="mt_seq" rows="3" cols="30" readonly>{{ external_data.sequences.mutant }}</textarea></td>
						<td class="w3-left-align"><em class="w3-small">Mutant type DNA sequence +/- 25 bp&nbsp;&nbsp;&nbsp;<button onclick="copy_text('mt_seq');" title="copy mutant sequence to clipboard"><i class="fa fa-clipboard" aria-hidden="true"></i></button></em></td>
					</tr>
				</tbody>
			</table><br/><br/>
			<div class="w3-row-padding">
				<div class="w3-col m4">
					<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" style="width:100%" id="litvar_data">
						Collecting LitVar data...<i class="fa fa-spinner w3-spin" style="font-size:24px"></i>
					</span>
				</div>
				{% if external_data.VCF.hg19.ncbiChr is not none %}
				<div class="w3-col m4" id="defgen_hg19">
					<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" title="generates a DEFGEN compliant csv file for the variant" style="width:100%" id="defgen_btn_hg19" onclick="defgen_export('hg19', '{{ external_data.variantId }}', '{{ url_for('ajax.defgen') }}', '{{ csrf_token() }}');">Export to DEFGEN (hg19)</span>
					<div class="w3-modal" id="defgen_modal_hg19" style="z-index:2002;"></div>
				</div>
				{% endif %}
				<div class="w3-col m4" id="defgen_hg38">
					<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" title="generates a DEFGEN compliant csv file for the variant" style="width:100%" id="defgen_btn_hg38" onclick="defgen_export('hg38', '{{ external_data.variantId }}', '{{ url_for('ajax.defgen') }}', '{{ csrf_token() }}');">Export to DEFGEN (hg38)</span>
					<div class="w3-modal" id="defgen_modal_hg38" style="z-index:2002;"></div>
				</div>
			</div>
		</div>
		<!--third section-->
		<hr />
		{{ section_title('frequencies', 'Population Frequencies and Databases') }}
		<div class="w3-responsive w3-container">
			<!--Position in transcript/protein-->
			<table class="w3-table w3-centered display compact" style="width:90%" id="population_table">
				<thead>
					<tr>
						<th class="w3-left-align">Features</th>
						<th class="w3-left-align" width="60%">Values</th>
						<th class="w3-left-align">Descriptions</th>
					</tr>
				</thead>
				<tbody>
					{% if external_data.VCF.hg19.ncbiChr is not none %}
					<tr><!--gnomAD exome-->
						<td class="w3-left-align"><a href={{ "%svariant/%s-%s-%s-%s?dataset=gnomad_r2_1"|format(urls.gnomad, external_data.VCF.chr, external_data.VCF.hg19.pos, external_data.VCF.hg19.ref, external_data.VCF.hg19.alt) }} target="_blank" title="Visit gnomAD for this variant">gnomAD</a> exome:</td>
						<td class="w3-left-align">{{ external_data.frequenciesDatabases.gnomADv2Exome }}</td>
						<td class="w3-left-align"><em class="w3-small"> v2.0.1 Exomes global MAF</em></td>
					</tr>
					<tr><!--gnomAD genome-->
						<td class="w3-left-align"><a href={{ "%svariant/%s-%s-%s-%s?dataset=gnomad_r2_1"|format(urls.gnomad, external_data.VCF.chr, external_data.VCF.hg19.pos, external_data.VCF.hg19.ref, external_data.VCF.hg19.alt) }} target="_blank" title="Visit gnomAD for this variant">gnomAD</a> genome:</td>
						<td class="w3-left-align">{{ external_data.frequenciesDatabases.gnomADv2Genome }}</td>
						<td class="w3-left-align"><em class="w3-small"> v2.0.1 Genomes global MAF</em></td>
					</tr>
					{% endif %}
					<tr><!--gnomAD v3-->
						<td class="w3-left-align"><a href={{ "%svariant/%s-%s-%s-%s?dataset=gnomad_r3"|format(urls.gnomad, external_data.VCF.chr, external_data.VCF.hg38.pos, external_data.VCF.hg38.ref, external_data.VCF.hg38.alt) }} target="_blank" title="Visit gnomAD for this variant">gnomAD</a> v3:</td>
						<td class="w3-left-align">{{ external_data.frequenciesDatabases.gnomADv3 }}</td>
						<td class="w3-left-align"><em class="w3-small"> v3 Genomes global MAF</em></td>
					</tr>
					<tr><!--dbSNP id-->
						<td class="w3-left-align">dbSNP rsid:</td>
						<td class="w3-left-align">
							{% if internal_data.frequenciesDatabases.dbSNPrsid  is not none %}
								<a href="{{ urls.ncbi_dbsnp }}rs{{ internal_data.frequenciesDatabases.dbSNPrsid }}" target="_blank" title="open dbSNP" id="dbsnp_id">rs{{ internal_data.frequenciesDatabases.dbSNPrsid }}</a>
							{% else %}
								No match in dbSNP {{ external_tools['dbSNP']['version'] }}
							{% endif %}
						</td>
						<td class="w3-left-align"><em class="w3-small">Identifier for NCBI dbSNP</em></td>
					</tr>
					<tr><!--clinvar-->
						<td class="w3-left-align">Clinvar:</td>
						<td class="w3-left-align">
							{% if external_data.frequenciesDatabases.clinvarId  is not none %}
								<a href={{ "%s/variation/%s/"|format(urls.ncbi_clinvar, external_data.frequenciesDatabases.clinvarId) }}  target="_blank" title="Visit Clinvar for this variant">{{ external_data.frequenciesDatabases.clinvarClinsig }}</a>
								{% if external_data.frequenciesDatabases.clinvarClinsigConf  is not none %}
									({{ external_data.frequenciesDatabases.clinvarClinsigConf }})
								{% endif %}
							{% else %}
								 {{ external_data.frequenciesDatabases.clinvarClinsig }}
							{%endif %}
						</td>
						<td class="w3-left-align"><em class="w3-small">
							{% if external_data.frequenciesDatabases.clinvarId  is not none %}
								<a href={{ "%s/variation/%s/"|format(urls.ncbi_clinvar, external_data.frequenciesDatabases.clinvarId) }}  target="_blank" title="Visit Clinvar for this variant">
							{% else %}
								<a href="{{ urls.ncbi_clinvar }}" target="_blank" title="Visit Clinvar">
							{%endif %}
							Clinvar</a> interpretation</em>
						</td>
					</tr>
					{% if external_data.gene.clingenCriteriaSpec %}
					<tr><!-- ClinGen Specification Registry -->
						<td class="w3-left-align">ClinGen criteria:</td>
						<td class="w3-left-align">
							<a href={{ "%sdoc/%s"|format(urls.clingen_gene_spec_web, external_data.gene.clingenCriteriaSpec) }}  target="_blank" title="Visit ClinGen to get specific ACMG criteria for this variant">{{ external_data.gene.clingenCriteriaSpec }}</a>
						</td>
						<td class="w3-left-align"><em class="w3-small">ClinGen ACMG specific rules for {{ external_data.gene.symbol }}</em></td>
					</tr>
					{% endif %}
					{% if external_data.positions.DNAType == 'substitution' and external_data.positions.segmentStartType == 'exon' and internal_data.nomenclatures.pName != '?' and internal_data.nomenclatures.pName != 'Met1?' %}
						<tr id="intervar_tr" onmouseout="$('#acmg_info').text('');"><!--intervar-->
							{# <td class="w3-left-align"><a href={{ "%shg19_updated.v.202107&chr=%s&pos=%s&ref=%s&alt=%s"|format(urls.intervar, external_data.VCF.chr, external_data.VCF.hg19.pos, external_data.VCF.hg19.ref, external_data.VCF.hg19.alt) }} target="_blank" title="Check and adjust ACMG class at intervar">Intervar</a>:</td> #}
							<td class="w3-left-align"><a href={{ "%shg38&chr=%s&pos=%s&ref=%s&alt=%s"|format(urls.intervar, external_data.VCF.chr, external_data.VCF.hg38.pos, external_data.VCF.hg38.ref, external_data.VCF.hg38.alt) }} target="_blank" title="Check and adjust ACMG class at intervar">hg38 InterVar</a>:</td>
							<td class="w3-left-align">
								<span id='intervar_data'>Asking Intervar...  <i class="fa fa-spinner w3-spin" style="font-size:24px"></i></span>
							</td>
							<td class="w3-left-align"><em class="w3-small">Semi-automated ACMG classification - click on the intervar link to adjust<br /> - passing the mouse over a criterion will display the definition</em>
							</td>
						</tr>
					{% endif %}
					<tr><!--LOVD matches-->
						<td class="w3-left-align" id="lovd_feature">LOVD matches:</td>
						<td class="w3-left-align">
							<span id='lovd_data'>Searching LOVD...  <i class="fa fa-spinner w3-spin" style="font-size:24px"></i></span>
						</td>
						<td class="w3-left-align" id="lovd_description"><em class="w3-small">LOVD match in public instances</em></td>
					</tr>
				</tbody>
			</table><br/><br/>
			<!--hidden values to trigger intervar/spliceaivisual ajax-->
			<span style="visibility:hidden" id="segment_type">{{ external_data.positions.segmentStartType }}</span>
			<span style="visibility:hidden" id="dna_type">{{ external_data.positions.DNAType }}</span>
			<span style="visibility:hidden" id="strand">{{ external_data.gene.strand }}</span>
		</div>
		<!--fourth section-->
		<hr />
		{{ section_title('predictions', 'Overall predictions') }}
		<div class="w3-responsive w3-container">
			<table class="w3-table w3-centered display compact" style="width:90%" id="prediction_table">
				<thead>
					<tr>
						<th class="w3-left-align">Features</th>
						<th class="w3-left-align" width="40%">Values</th>
						<th class="w3-left-align">Descriptions</th>
					</tr>
				</thead>
				<tbody>
					{% if g.user['academic'] or not g.user %}
						<tr><!--CADD raw-->
							<!--https://cadd.gs.washington.edu/snv/GRCh37-v1.6_inclAnno/22:43451447_C_A-->
							<td class="w3-left-align"><a href="{{ urls.cadd }}" target="_blank" title="Visit CADD">CADD</a> raw:</td>
							{% if internal_data.noMatch.cadd is none %}
								<td class="w3-left-align">{{ external_data.overallPredictions.caddRaw }}</td>
							{% else %}
								<td class="w3-left-align">{{ internal_data.noMatch.cadd }}</td>
							{% endif %}
							<td class="w3-left-align"><em class="w3-small">[-6.41;35.5] The higher the less likely to be observed</em></td>
						</tr>
						<tr><!--CADD phred-->
							<td class="w3-left-align"><a href="{{ urls.cadd }}" target="_blank" title="Visit CADD">CADD</a> phred:</td>
							{% if internal_data.noMatch.cadd is none %}
								<td class="w3-left-align">{{ external_data.overallPredictions.caddPhred }}</td>
							{% else %}
								<td class="w3-left-align">{{ internal_data.noMatch.cadd }}</td>
							{% endif %}
							<td class="w3-left-align"><em class="w3-small">Phred-like scaling of raw score</em></td>
						</tr>
					{% endif %}
					{% if internal_data.noMatch.eigen is none or external_data.overallPredictions.eigenRaw is not none %}
						<tr><!--Eigen raw-->
							<td class="w3-left-align"><a href="{{ urls.eigen }}" target="_blank" title="Visit Eigen">Eigen</a> raw:</td>
							{% if internal_data.noMatch.eigen is none %}
								<td class="w3-left-align">{{ external_data.overallPredictions.eigenRaw }}</td>
							{% else %}
								<td class="w3-left-align">{{ internal_data.noMatch.eigen }}</td>
							{% endif %}
							<td class="w3-left-align"><em class="w3-small">[-3.33;6.84] The higher the less likely to be observed</em></td>
						</tr>
						<tr><!--Eigen phred-->
							<td class="w3-left-align"><a href="{{ urls.eigen }}" target="_blank" title="Visit Eigen">Eigen</a> phred:</td>
							{% if internal_data.noMatch.eigen is none %}
								<td class="w3-left-align">{{ external_data.overallPredictions.eigenPhred }}</td>
							{% else %}
								<td class="w3-left-align">{{ internal_data.noMatch.eigen }}</td>
							{% endif %}
							<td class="w3-left-align"><em class="w3-small">Phred-like scaling of raw score</em></td>
						</tr>
					{% endif %}
					<tr><!--MPA score-->
						<td class="w3-left-align"><a href="{{ urls.mpa }}" target="_blank" title="VMPA GitHub">MPA</a> score:</td>
							<td class="w3-left-align">{{ span_color(external_data.overallPredictions.mpaScore, internal_data.overallPredictions.mpaColor) }}</td>
						<td class="w3-left-align"><em class="w3-small">Raw score [0;10], 10: high impact</em></td>
					</tr>
					<tr><!--MPA impact-->
						<td class="w3-left-align"><a href="{{ urls.mpa }}" target="_blank" title="MPA GitHub">MPA</a> impact:</td>
							<td class="w3-left-align">{{ span_color(external_data.overallPredictions.mpaImpact.capitalize(), internal_data.overallPredictions.mpaColor) }}</td>
						<td class="w3-left-align"><em class="w3-small">Impact type</em></td>
					</tr>
				</tbody>
			</table><br/><br/>
		</div>
		<!--fifth section-->
		<hr />
		{{ section_title('splicing', 'Splicing predictions') }}
		<div class="w3-responsive w3-container">
			{% if (external_data.positions.segmentStartType == 'exon' or (external_data.positions.segmentStartType == 'intron' and internal_data.positions.distFromExon < 101)) and external_data.positions.segmentStartType == external_data.positions.segmentEndType and external_data.positions.size < 50 %}
				<!-- Exonic Splicing context -->
				<br />{{ info("The exonic splicing context of the variant including natural splice sites scores is summarized in the graph below:") }}
				<div class="w3-container w3-center">
					<canvas class="exonic_canvas" width="600" height="150" id="segment_drawing">Your web browser does not understand HTML5. In order to display the small graph, you must use a more recent one.</canvas>
				</div>
				{% autoescape false %}
					<script type="text/javascript">
						var canvas = document.getElementById("segment_drawing");
						var context = canvas.getContext("2d");
						context.fillStyle = "#000000";
						context.font = "bold 14px sans-serif";
						context.strokeStyle = "#2196F3";
						{% if internal_data.canvas.preceedingSegmentType is not none %}
							context.fillText("{{ internal_data.canvas.preceedingSegmentType }} {{ internal_data.canvas.preceedingSegmentNumber }}", 100, 25);
						{% endif %}
						{% if internal_data.canvas.posExonCanvas is not none %}
							context.fillText("c.{{ internal_data.nomenclatures.cName }}", {{ internal_data.canvas.posExonCanvas }}, 20);
						{% else %}
							context.fillText("c.{{ internal_data.nomenclatures.cName }}", {{ internal_data.canvas.posIntronCanvas }}, 45);
						{% endif %}
						{% if internal_data.canvas.followingSegmentType is not none %}
							context.fillText("{{ internal_data.canvas.followingSegmentType }} {{ internal_data.canvas.followingSegmentNumber }}", 500, 25);
						{% endif %}
						{% if internal_data.canvas.preceedingSegmentNumber != 'UTR' and internal_data.canvas.preceedingSegmentNumber is not none %}
							context.fillText("3'ss MaxEnt: {{ internal_data.splicingPredictions.nat3ssScore }}", 150, 125);
							context.fillText("{{ internal_data.splicingPredictions.nat3ssSeq }}", 100, 140);
						{% endif %}
						{% if internal_data.canvas.followingSegmentNumber != 'UTR' and internal_data.canvas.followingSegmentNumber is not none %}
							context.fillText("5'ss MaxEnt: {{ internal_data.splicingPredictions.nat5ssScore }}", 400, 125);
							context.fillText("{{ internal_data.splicingPredictions.nat5ssSeq }}", 400, 140);
						{% endif %}
						context.moveTo(100,75);
						context.lineTo(200,75);
						context.stroke();
						{% if internal_data.canvas.posExonCanvas is not none %}
							context.moveTo({{ internal_data.canvas.posExonCanvas }},20);
							context.lineTo({{ internal_data.canvas.posExonCanvas }},20+30);
						{% else %}
							context.moveTo({{ internal_data.canvas.posIntronCanvas }},45);
							context.lineTo({{ internal_data.canvas.posIntronCanvas }},45+30);
						{% endif %}
						context.stroke();
						context.fillStyle = "#fdf5e6"
						context.fillRect(200,50,200,50);
						context.strokeRect(200,50,200,50);
						context.fillStyle = "#000000";
						{% if external_data.positions.segmentStartType == 'exon' %}
							context.fillText("Exon {{ external_data.positions.segmentStartNumber }} ({{ internal_data.canvas.segmentSize }} bp)", 250, 75);
						{% else %}
							context.fillText("Exon {{ internal_data.positions.neighbourExonNumber.neighb_exon_number }}", 250, 75);
						{% endif %}
						context.moveTo(400,75);
						context.lineTo(500,75);
						context.stroke();
					</script>
				{% endautoescape %}
			{% endif %}
			<!-- MaxEntScan 5'ss -->
			{{ info('MaxEntScan scores are presented in the two following tables. Selected scores have: <ul><li>a |variation| > 15% and</li><li>a raw score for mutant or wild-type of at least 3</li></ul> The upper sequence shows the variation site in red, and the lower sequence the putative splice site considered by <a href ="http://hollywood.mit.edu/burgelab/maxent/Xmaxentscan_scoreseq.html" target="_blank">MaxEntScan</a> (putative introns are shown as lower case and exons as upper case).') }}
			<br />
			<table class="w3-table w3-centered display compact" style="width:90%;" id="maxent5ss_table">
				<caption>MaxEntScan 5'ss scores</caption>
				<thead>
					<tr>
						<th class="w3-left-align">Wild-type sequence</th>
						<th class="w3-left-align">Score</th>
						<th class="w3-left-align">Mutant sequence</th>
						<th class="w3-left-align">Score</th>
						<th class="w3-left-align">Variation(%)</th>
					</tr>
				</thead>
				<tbody>
					{% if external_data.splicingPredictions.mes5 is not none %}
						{% autoescape false %}
							{% for score in external_data.splicingPredictions.mes5 %}
								<tr>
									<td>... {{ external_data.splicingPredictions.mes5[score][5] }}...<br />... {{ external_data.splicingPredictions.mes5[score][6] }} ...</td>
									<td style="vertical-align:middle;">{{ external_data.splicingPredictions.mes5[score][1] }}</td>
									<td>... {{ external_data.splicingPredictions.mes5[score][7] }}...<br />... {{ external_data.splicingPredictions.mes5[score][8] }} ...</td>
									<td style="vertical-align:middle;">{{ external_data.splicingPredictions.mes5[score][3] }}</td>
									{% if external_data.splicingPredictions.mes5[score][4] > 0 %}
										<td style="vertical-align:middle;">+{{ external_data.splicingPredictions.mes5[score][4] }}</td>
									{% else %}
										<td style="vertical-align:middle;">{{ external_data.splicingPredictions.mes5[score][4] }}</td>
									{% endif %}
								</tr>
							{% endfor %}
						{% endautoescape %}
					{% else %}
						<tr><td></td><td></td><td>No significant MaxEnt 5'ss scores found.</td><td></td><td></td></tr>
					{% endif %}
				</tbody>
			</table><br/><br/>
			<!-- MaxEntScan 3'ss -->
			<table class="w3-table w3-centered display compact" style="width:90%" id="maxent3ss_table">
				<caption>MaxEntScan 3'ss scores</caption>
				<thead>
					<tr>
						<th class="w3-left-align">Wild-type sequence</th>
						<th class="w3-left-align">Score</th>
						<th class="w3-left-align">Mutant sequence</th>
						<th class="w3-left-align">Score</th>
						<th class="w3-left-align">Variation(%)</th>
					</tr>
				</thead>
				<tbody>
					{% if external_data.splicingPredictions.mes3 is not none and external_data.splicingPredictions.mes3 != 'Not performed' %}
						{% autoescape false %}
							{% for score in external_data.splicingPredictions.mes3 %}
								<tr>
									<td>... {{ external_data.splicingPredictions.mes3[score][5] }}...<br />... {{ external_data.splicingPredictions.mes3[score][6] }} ...</td>
									<td style="vertical-align:middle;">{{ external_data.splicingPredictions.mes3[score][1] }}</td>
									<td>... {{ external_data.splicingPredictions.mes3[score][7] }}...<br />... {{ external_data.splicingPredictions.mes3[score][8] }} ...</td>
									<td style="vertical-align:middle;">{{ external_data.splicingPredictions.mes3[score][3] }}</td>
									{% if external_data.splicingPredictions.mes3[score][4] > 0 %}
										<td style="vertical-align:middle;">+{{ external_data.splicingPredictions.mes3[score][4] }}</td>
									{% else %}
										<td style="vertical-align:middle;">{{ external_data.splicingPredictions.mes3[score][4] }}</td>
									{% endif %}
								</tr>
							{% endfor %}
						{% endautoescape %}
					{% elif external_data.splicingPredictions.mes3 == 'Not performed' %}
						<tr><td></td><td></td><td>No MaxEnt 3'ss score performed (exonic variant far form 3'ss).</td><td></td><td></td></tr>
					{% else %}
						<tr><td></td><td></td><td>No significant MaxEnt 3'ss scores found.</td><td></td><td></td></tr>
					{% endif %}
				</tbody>
			</table><br/><br/>
			<!-- SPiP -->
			{{ info('<a href="%s" target="_blank">SPiP</a> results and predictions:' | format(urls.spip)) }}
			<div id="spip">{{ info('MobiDetails is running SPiPv2 (may last up to 20s)... <i class="fa fa-spinner w3-spin" style="font-size:24px"></i>') }}</div>
			{% if external_data.positions.DNAType != 'indel' or (internal_data.positions.insSize is not none and internal_data.positions.insSize < 2) %}
				<!--dbscSNV / spliceAI-->
				{% if external_data.positions.DNAType != 'indel' %}
					{{ info('This last table presents dbscSNV and spliceAi scores (max distance: 50 bp), when available. You can also (except for delins) run spliceAI via the <a href="%s" target="_blank">spliceAIlookup API</a> using a distance of 500 bp. This may take a while especially for genes with multiple transcripts but may help you identify new deep intronic variants. This also gives the opportunity to get spliceAi scores for large insertions or deletions. ')|format(urls.spliceai_lookup) }}
				{% endif %}
				{# {% if external_data.positions.DNAType == 'substitution' or external_data.splicingPredictions.spliceai_DS_AG is not none %} #}
				<table class="w3-table w3-centered display compact" style="width:90%;" id="splicing_table">
					<thead>
						<tr>
							<th class="w3-left-align">Features</th>
							<th class="w3-left-align" width="40%">Values</th>
							<th class="w3-left-align">Descriptions</th>
						</tr>
					</thead>
					<tbody>
						{% if external_data.positions.DNAType == 'substitution' %}
							<tr><!--dbscSNV ADA-->
								<td class="w3-left-align"><a href="{{ urls.dbscsnv }}" target="_blank" title="Visit dbscSNV">dbscSNV</a> ADA:</td>
								<td class="w3-left-align">{{ span_color(external_data.splicingPredictions.dbscSNVADA, internal_data.splicingPredictions.dbscSNVADAColor) }}</td>
								<td class="w3-left-align"><em class="w3-small">Raw score (0-1), threshold &gt; {{ thresholds.dbscsnv }} for impact</em></td>
							</tr>
							<tr><!--dbscSNV RF-->
								<td class="w3-left-align"><a href="{{ urls.dbscsnv }}" target="_blank" title="Visit dbscSNV">dbscSNV</a> RF:</td>
								<td class="w3-left-align">{{ span_color(external_data.splicingPredictions.dbscSNVRF, internal_data.splicingPredictions.dbscSNVRFColor) }}</td>
								<td class="w3-left-align"><em class="w3-small">Raw score (0-1), threshold &gt; {{ thresholds.dbscsnv }} for impact</em></td>
							</tr>
						{% endif %}
						{% if g.user['academic'] or not g.user %}
							{% if external_data.splicingPredictions.spliceai_DS_AG is not none %}
								<tr id='spliceai_ag_50_tr'><!--spliceAI AG-->
									<td class="w3-left-align"><a href="{{ urls.spliceai }}" target="_blank" title="Visit spliceAI">spliceAI</a> AG:</td>
									<td class="w3-left-align">{{ span_color(external_data.splicingPredictions.spliceai_DS_AG, internal_data.splicingPredictions.spliceai_DS_AG_color) }} ({{ external_data.splicingPredictions.spliceai_DP_AG }})</td>
									<td class="w3-left-align"><em class="w3-small">Acceptor Gain, 0-1 (distance in bp), thresholds &gt; {{ "%s|%s|%s"|format(thresholds.spliceai_min, thresholds.spliceai_mid, thresholds.spliceai_max) }} for impact</em></td>
								</tr>
								<tr id='spliceai_al_50_tr'><!--spliceAI AL-->
									<td class="w3-left-align"><a href="{{ urls.spliceai }}" target="_blank" title="Visit spliceAI">spliceAI</a> AL:</td>
									<td class="w3-left-align">{{ span_color(external_data.splicingPredictions.spliceai_DS_AL, internal_data.splicingPredictions.spliceai_DS_AL_color) }} ({{ external_data.splicingPredictions.spliceai_DP_AL }})</td>
									<td class="w3-left-align"><em class="w3-small">Acceptor Loss, 0-1 (distance in bp), thresholds &gt; {{ "%s|%s|%s"|format(thresholds.spliceai_min, thresholds.spliceai_mid, thresholds.spliceai_max) }} for impact</em></td>
								</tr>
								<tr id='spliceai_dg_50_tr'><!--spliceAI DG-->
									<td class="w3-left-align"><a href="{{ urls.spliceai }}" target="_blank" title="Visit spliceAI">spliceAI</a> DG:</td>
									<td class="w3-left-align">{{ span_color(external_data.splicingPredictions.spliceai_DS_DG, internal_data.splicingPredictions.spliceai_DS_DG_color) }} ({{ external_data.splicingPredictions.spliceai_DP_DG }})</td>
									<td class="w3-left-align"><em class="w3-small">Donor Gain, 0-1 (distance in bp), thresholds &gt; {{ "%s|%s|%s"|format(thresholds.spliceai_min, thresholds.spliceai_mid, thresholds.spliceai_max) }} for impact</em></td>
								</tr>
								<tr id='spliceai_dl_50_tr'><!--spliceAI DL-->
									<td class="w3-left-align"><a href="{{ urls.spliceai }}" target="_blank" title="Visit spliceAI">spliceAI</a> DL:</td>
									<td class="w3-left-align">{{ span_color(external_data.splicingPredictions.spliceai_DS_DL, internal_data.splicingPredictions.spliceai_DS_DL_color) }} ({{ external_data.splicingPredictions.spliceai_DP_DL }})</td>
									<td class="w3-left-align"><em class="w3-small">Donor Loss, 0-1 (distance in bp), thresholds &gt; {{ "%s|%s|%s"|format(thresholds.spliceai_min, thresholds.spliceai_mid, thresholds.spliceai_max) }} for impact</em></td>
								</tr>
							{% elif internal_data.noMatch.spliceai is not none %}
								<tr>
									<td class="w3-left-align"><a href="{{ urls.spliceai }}" target="_blank" title="Visit spliceAI">spliceAI</a>:</td>
									<td class="w3-left-align">{{ internal_data.noMatch.spliceai }}</td>
									<td class="w3-left-align"><em class="w3-small">Donor Loss, 0-1 (distance in bp), thresholds &gt; {{ "%s|%s|%s"|format(thresholds.spliceai_min, thresholds.spliceai_mid, thresholds.spliceai_max) }} for impact</em></td>
								</tr>
							{% endif %}
						{% endif %}
						<tr id="spliceai_lookup_ag_tr" style="display:none;"><!--spliceAI lookup AG-->
							<td class="w3-left-align"><a href="{{ urls.spliceai_lookup }}" target="_blank" title="Visit spliceAI lookup">spliceAI lookup (500)</a> AG:</td>
							<td class="w3-left-align" id="spliceai_lookup_ag">SpliceAi 500 was not run</td>
							<td class="w3-left-align"><em class="w3-small">Acceptor Gain, 0-1 (distance in bp), thresholds &gt; {{ "%s|%s|%s"|format(thresholds.spliceai_min, thresholds.spliceai_mid, thresholds.spliceai_max) }} for impact</em></td>
						</tr>
						<tr id="spliceai_lookup_al_tr" style="display:none;"><!--spliceAI lookup AL-->
							<td class="w3-left-align"><a href="{{ urls.spliceai_lookup }}" target="_blank" title="Visit spliceAI lookup">spliceAI lookup (500)</a> AL:</td>
							<td class="w3-left-align" id="spliceai_lookup_al">SpliceAi 500 was not run</td>
							<td class="w3-left-align"><em class="w3-small">Acceptor Loss, 0-1 (distance in bp), thresholds &gt; {{ "%s|%s|%s"|format(thresholds.spliceai_min, thresholds.spliceai_mid, thresholds.spliceai_max) }} for impact</em></td>
						</tr>
						<tr id="spliceai_lookup_dg_tr" style="display:none;"><!--spliceAI lookup DG-->
							<td class="w3-left-align"><a href="{{ urls.spliceai_lookup }}" target="_blank" title="Visit spliceAI lookup">spliceAI lookup (500)</a> DG:</td>
							<td class="w3-left-align" id="spliceai_lookup_dg">SpliceAi 500 was not run</td>
							<td class="w3-left-align"><em class="w3-small">Donor Gain, 0-1 (distance in bp), thresholds &gt; {{ "%s|%s|%s"|format(thresholds.spliceai_min, thresholds.spliceai_mid, thresholds.spliceai_max) }} for impact</em></td>
						</tr>
						<tr id="spliceai_lookup_dl_tr" style="display:none;"><!--spliceAI lookup DL-->
							<td class="w3-left-align"><a href="{{ urls.spliceai_lookup }}" target="_blank" title="Visit spliceAI lookup">spliceAI lookup (500)</a> DL:</td>
							<td class="w3-left-align" id="spliceai_lookup_dl">SpliceAi 500 was not run</td>
							<td class="w3-left-align"><em class="w3-small">Donor Loss, 0-1 (distance in bp), thresholds &gt; {{ "%s|%s|%s"|format(thresholds.spliceai_min, thresholds.spliceai_mid, thresholds.spliceai_max) }} for impact</em></td>
						</tr>
					</tbody>
				</table>
			{% endif %}
				<br/><br/>
				<div class="w3-row-padding">
			{% if external_data.positions.DNAType != 'indel' or (internal_data.positions.insSize is not none and internal_data.positions.insSize < 2) %}
					{% if external_data.splicingPredictions.spliceai_DS_AG is not none %}
						<!--Splicing radar chart-->
						<div class="w3-col m4">
							<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="$('#radar_splicing_modal').show();" title="Display a radar chart of splicing scores" style="width:100%">Show radar chart</span>
						</div>
					{% endif %}
					<!--spliceai lookup button-->
					{% if external_data.positions.DNAType != 'indel' %}
						<div class="w3-col m4" id="spliceai_lookup_button">
							<button id="spliceai_lookup_button_span" class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="$('html').css('cursor', 'progress');$('#spliceai_lookup_button_span').addClass('w3-disabled');this.disabled=true;spliceai_lookup('{{ url_for('ajax.spliceai_lookup') }}', '{{ csrf_token() }}');$('#spliceai_lookup_wheel').html('<i class=\'fa fa-spinner w3-spin\' style=\'font-size:24px\'></i>');" title="Run SpliceAI lookup API with a distance of 500 bp instead of the default 50 bp - may be useful to identify e.g. deep intronic variant which trigger exonisation" style="width:100%">Run SpliceAI 500&nbsp;&nbsp;<span id="spliceai_lookup_wheel"></span></button>
						</div>
					{% endif %}
			{% endif %}
					<!--SpliceAI-visual full transcript button-->
					<div class="w3-col m4" id="spliceai_visual_full_gene" style="display:none;">
						<button id="spliceai_visual_full_gene_span" class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="$('html').css('cursor', 'progress');this.disabled=true;spliceaivisual('{{ url_for('ajax.spliceaivisual') }}', '{{ url_for('static', filename='') }}', 'spliceaivisual_full_button', '{{ csrf_token() }}');$('#spliceai_visual_full_wheel').html('<i class=\'fa fa-spinner w3-spin\' style=\'font-size:24px\'></i>');" title="Run SpliceAI-visual for the entire mutant transcript - may take several minutes to complete (e.g. 5 minutes for a 1Mb transcript)" style="width:100%">SpliceAI-visual (full transcript)<span id="spliceai_visual_full_wheel"></span></button>
					</div>
				</div>
				<br />
			{% if external_data.positions.DNAType != 'indel' or (internal_data.positions.insSize is not none and internal_data.positions.insSize < 2) %}
				{% if external_data.splicingPredictions.spliceai_DS_AG is not none %}
					<div id="radar_splicing_modal" class="w3-modal w3-animate-opacity" style="display:block;z-index:2002;">
						<div class="w3-modal-content w3-card-4 w3-display-middle w3-animate-opacity w3-mobile">
							<header class="w3-container w3-blue">
								<span class="w3-button w3-display-topright w3-xlarge" onclick="$('#radar_splicing_modal').hide();">X</span>
								<br /><span class="w3-xlarge">Radar view of splicing predictors</span><br />
								<p>Values are normalised (0-1), 0 being the less damaging and 1 the most for each predictor.</p>
							</header>
							<div class="w3-container">
								<div class="chart-container" style="position: relative;">
									<canvas id="splicing_radar"><p>Radar Chart for splicing predictions</p></canvas>
								</div>
							</div>
						</div>
					</div>
					<script>
						Chart.defaults.global.defaultFontSize=16;
						var point_font_size = 16
						var wi = $(window).width();
						if (wi < 400) {
							Chart.defaults.global.defaultFontSize=12;
							point_font_size = 12;
						}
						{% autoescape false %}
							var myRadarChart = new Chart($("#splicing_radar"), {
								type: 'radar',
								data: {
									//labels: ['dbscSNV ADA', 'dbscSNV RF', 'SpliceAI Acc Gain', 'SpliceAI Acc Loss', 'SpliceAI Donor Gain', 'SpliceAI Donor Loss'],
									labels: {{ internal_data.splicingPredictions.splicingRadarLabels }},
									datasets: [{
										label:'Splicing predictors',
										data: {{ internal_data.splicingPredictions.splicingRadarValues }},
										fill: true,
										backgroundColor:'rgb(255, 99, 132, 0.2)',
										borderColor:'rgb(255, 99, 132)',
										pointBackgroundColor:'rgb(255, 99, 132)',
										pointBorderColor:'#fff',
										pointHoverBackgroundColor:'#fff',
										pointHoverBorderColor:'rgb(255, 99, 132)'
									}]
								},
								options:{elements:{line:{tension:0,borderWidth:3}},scale:{pointLabels:{fontSize:point_font_size},ticks:{beginAtZero:true,min:0,max:1,stepSize:0.2}},multiTooltipTemplate : "<%=label%> : <%=value%>",onAnimationComplete: "done"}
							});
						{% endautoescape %}
					</script>
				{% endif %}
			{% endif %}
			<!--spliceaivisual in igv.js-->
			<br />
			<div id="igv_desc" style="display:none">
				{{ info('SpliceAI-visual bedgraphs are displayed below as tracks in igv.js.<ul><li>Green and red highlights show the variation site on the wild-type and mutant track, respectively</li><li>Positive orange bars represent predicted acceptor sites ([0;1]), and negative blue bars donor sites ([-1;0]), the spliceai raw score being the absolute value.</li><li>The first track is the spliceai prediction for the entire MANE transcript.</li><li>The second track is the prediction for the variant, +/- 10kb inside the transcript (i.e. not considering intergenic nucleotides).</li><li>The "inserted nucleotides" track is used to display the scores of the nucleotides being part of the insertion (empty if no insertion)</li><li>You can download the <a href="{0}" target="_blank">Wild-type</a> and <a href="{1}" target="_blank">mutant</a> <span id="full_transcript_download"></span> bedgraphs in order to load them in a third-party application such as your own Desktop IGV or the UCSC genome browser.</li><li><strong>Important notice: </strong>if your intronic variant is located more than 10kb from one of the two flanking exons, you are advised to click on the <strong>"SpliceAI-visual full transcript"</strong> button at the top of this text. Depending on the size of your transcript, within 30s-10 minutes (for the DMD gene, the bad news), <strong>the whole transcript prediction including the variant will appear as a new track in the igv.js browser</strong>, allowing a more accurate prediction for your variant. <strong>The good news being that if is has already been computed, it will load automatically.</strong></li></ul>'.format(url_for("static", filename="resources/spliceai/transcripts/{0}.bedGraph".format(external_data.gene.RefSeqTranscript)), url_for("static", filename="resources/spliceai/variants/{0}.bedGraph".format(external_data.variantId)))) }}
			</div>
			<div id="igv_com">
				{{ info('MobiDetails is running spliceai to bring you SpliceAI-visual... <i class="fa fa-spinner w3-spin" style="font-size:24px"></i>') }}
			</div>
			<div id="igv_div">
			</div>
			<br />
		</div>
		<!--alt sixth section-->
		{% if external_data.positions.DNAType == 'substitution' and internal_data.noMatch.dbmts is none and external_data.miRNATargetSitesPredictions.mirandaCategory is not none %}
			<hr />
			{{ section_title('dbmts', 'miRNA target sites predictions') }}
			<div class="w3-responsive w3-container">
				<!--dbMTS-->
				 {{ info("The table below comes form <a href='http://www.liulab.science/dbmts.html' target='_blank'>dbMTS{}</a>. It focuses on miRNA binding capabilities of the wild-type and mutant sequences and on the differences observed using 3 different algorithms, namely <a href='http://www.microrna.org/microrna/getGeneForm.do' target='_blank'>miRanda</a>, <a href='http://www.rnaiweb.com/cgi-bin/web/jump.cgi?ID=56' target='_blank'>TargetScan</a> and <a href='https://bibiserv.cebitec.uni-bielefeld.de/rnahybrid/' target='_blank'>RNAHybrid</a>. Each predictor may define three categories of functional change (see Descriptions). To correctly assess the Max Diff column, one needs to know the theoretical max differences for each tool: miRanda (216), TargetScan (11.07), RNAHybrid (60.8). Concerning best scores, for miRanda and RNAhybrid, the best score is the miRNA-target pair that has the highest prediction score. For TargetScan, the best score is the miRNA-target pair that has the largest negative prediction score.".format(external_tools['dbMTS']['version'])) }}
				<table class="w3-table w3-centered display compact" style="width:90%" id="dbmts_table">
					<thead>
						<tr>
							<th class="w3-left-align">Features</th>
							<th class="w3-left-align">miRanda</th>
							<th class="w3-left-align">TargetScan v7.0</th>
							<th class="w3-left-align">RNAhybrid</th>
							<th class="w3-left-align">Descriptions</th>
						</tr>
					</thead>
					<tbody>
						<tr><!--category-->
							<td class="w3-left-align">Category:</td>
							<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.mirandaCategory }}</td>
							<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.targetScanCategory }}</td>
							<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.RNAHybridCategory }}</td>
							<td class="w3-left-align"><em class="w3-small">Predicted functional category of this variant according to each predictor. 'S': substitution of regulating miRNAs. 'L': loss of all miRNAs. 'G' gain.</em></td>
						</tr>
						<tr><!--rankscore-->
							<td class="w3-left-align">Rank Score:</td>
							<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.mirandaRankScore }}</td>
							<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.targetScanRankScore }}</td>
							<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.RNAHybridRankScore }}</td>
							<td class="w3-left-align"><em class="w3-small">Rank of the max_dif score among all max_dif scores within the same category [0;1]. The closer to 1 the more likely to highlight a significant change.</em></td>
						</tr>
						<tr><!--maxdiff-->
							<td class="w3-left-align">Max Diff:</td>
							<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.mirandaMaxDiff }}</td>
							<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.targetScanMaxDiff }}</td>
							<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.RNAHybridMaxDiff }}</td>
							<td class="w3-left-align"><em class="w3-small">The absolute value of the maximum potential difference between all the scores (ref and alt).</em></td>
						</tr>
						{% autoescape false %}
							<tr><!--refbestmir-->
								<td class="w3-left-align">Ref Best Mir:</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.mirandaRefBestMir }}</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.targetScanRefBestMir }}</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.RNAHybridRefBestMir }}</td>
								<td class="w3-left-align"><em class="w3-small">miRNA corresponding to the best score of the reference (wild-type) sequence.</em></td>
							</tr>
							<tr><!--refbestscore-->
								<td class="w3-left-align">Ref Best Score:</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.mirandaRefBestScore }}</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.targetScanRefBestScore }}</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.RNAHybridRefBestScore }}</td>
								<td class="w3-left-align"><em class="w3-small">Best score for the reference sequence for each predictor.</em></td>
							</tr>
							<tr><!--altbestmir-->
								<td class="w3-left-align">Alt Best Mir:</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.mirandaAltBestMir }}</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.targetScanAltBestMir }}</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.RNAHybridAltBestMir }}</td>
								<td class="w3-left-align"><em class="w3-small">miRNA corresponding to the best score of the alternative (mutant) sequence.</em></td>
							</tr>
							<tr><!--altbestscore-->
								<td class="w3-left-align">Alt Best Score:</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.mirandaAltBestScore }}</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.targetScanAltBestScore }}</td>
								<td class="w3-left-align">{{ external_data.miRNATargetSitesPredictions.RNAHybridAltBestScore }}</td>
								<td class="w3-left-align"><em class="w3-small">Best score for the alternative sequence for each predictor.</em></td>
							</tr>
						{% endautoescape %}
					</tbody>
				</table>
			</div>
		{% elif internal_data.noMatch.dbmts is not none %}
			{{ info(internal_data.noMatch.dbmts) }}
		{% endif %}
		<!--missense section-->
		{% if external_data.positions.proteinType == 'missense' and internal_data.noMatch.dbnsfp is none %}
			<hr />
			{{ section_title('missense', 'Missense predictions: p.(%s)'|format(internal_data.nomenclatures.pName)) }}
			<div class="w3-responsive w3-container">
				<!--dbscSNV-->
				<table class="w3-table w3-centered display compact" style="width:90%" id="missense_table">
					<thead>
						<tr>
							<th class="w3-left-align">Features</th>
							<th class="w3-left-align">Values</th>
							<th class="w3-left-align">Prediction</th>
							<th class="w3-left-align">Descriptions</th>
						</tr>
					</thead>
					<tbody>
						<tr><!--SIFT-->
							<td class="w3-left-align"><a href="{{ urls.sift }}" target="_blank" title="Visit SIFT 4G">SIFT</a>:</td>
							<td class="w3-left-align">{{ span_color(external_data.missensePredictions.siftScore, internal_data.missensePredictions.siftColor) }}{{ internal_data.missensePredictions.siftStar }}</td>
							<td class="w3-left-align">{{ span_color(external_data.missensePredictions.siftPred, internal_data.missensePredictions.siftColor) }}</td>
							<td class="w3-left-align"><em class="w3-small">Threshold &lt; 0.05 for Damaging - single score</em></td>
						</tr>
						{% if g.user['academic'] or not g.user %}
							<tr><!--PH2 HDIV-->
								<td class="w3-left-align"><a href="{{ urls.pph2 }}" target="_blank" title="Visit Polyphen">Polyphen 2</a> HumDiv:</td>
								<td class="w3-left-align">{{ span_color(external_data.missensePredictions.polyphen2HdivScore, internal_data.missensePredictions.polyphen2HdivColor) }}{{ internal_data.missensePredictions.polyphen2HdivStar }}</td>
								<td class="w3-left-align">{{ span_color(external_data.missensePredictions.polyphen2HdivPred, internal_data.missensePredictions.polyphen2HdivColor) }}</td>
								<td class="w3-left-align"><em class="w3-small">Thresholds &gt; {{ "%s|%s"|format(thresholds.pph2_hdiv_mid, thresholds.pph2_hdiv_max) }} for Possibly and Probably Damaging - single score</em></td>
							</tr>
							<tr><!--PH2 HVAR-->
								<td class="w3-left-align"><a href="{{ urls.pph2 }}" target="_blank" title="Visit Polyphen">Polyphen 2</a> HumVar:</td>
								<td class="w3-left-align">{{ span_color(external_data.missensePredictions.polyphen2HvarScore, internal_data.missensePredictions.polyphen2HvarColor) }}{{ internal_data.missensePredictions.polyphen2HvarStar }}</td>
								<td class="w3-left-align">{{ span_color(external_data.missensePredictions.polyphen2HvarPred, internal_data.missensePredictions.polyphen2HvarColor) }}</td>
								<td class="w3-left-align"><em class="w3-small">Thresholds &gt; {{ "%s|%s"|format(thresholds.pph2_hvar_mid, thresholds.pph2_hvar_max) }} for Possibly and Probably Damaging - single score</em></td>
							</tr>
						{% endif %}
							<tr><!--FATHMM-->
								<td class="w3-left-align"><a href="{{ urls.fathmm }}" target="_blank" title="Visit fathmm">Fathmm</a>:</td>
								<td class="w3-left-align">{{ span_color(external_data.missensePredictions.fathmmScore, internal_data.missensePredictions.fathmmColor) }}{{ internal_data.missensePredictions.fathmmstar }}</td>
								<td class="w3-left-align">{{ span_color(external_data.missensePredictions.fathmmPred, internal_data.missensePredictions.fathmmColor) }}</td>
								<td class="w3-left-align"><em class="w3-small">Threshold &le; {{ thresholds.fathmm }} for Damaging - single score</em></td>
							</tr>
						{% if g.user['academic'] or not g.user %}
							<tr><!--REVEL-->
								<td class="w3-left-align"><a href="{{ urls.revel }}" target="_blank" title="Visit REVEL">REVEL</a>:</td>
								<td class="w3-left-align">{{ span_color(external_data.missensePredictions.revelScore, internal_data.missensePredictions.revelColor) }}</td>
								<td class="w3-left-align">{{ span_color(external_data.missensePredictions.revelPred, internal_data.missensePredictions.revelColor) }}</td>
								<td class="w3-left-align"><em class="w3-small">Thresholds 0.2|0.5 for Benign, Uncertain, Damaging - meta score</em></td>
							</tr>
							<tr><!--ClinPred-->
								<td class="w3-left-align"><a href="{{ urls.clinpred }}" target="_blank" title="Visit ClinPred">ClinPred</a>:</td>
								<td class="w3-left-align">{{ span_color(external_data.missensePredictions.clinpredScore, internal_data.missensePredictions.clinpredColor) }}{{ internal_data.missensePredictions.clinpredStar }}</td>
								<td class="w3-left-align">{{ span_color(external_data.missensePredictions.clinpredPred, internal_data.missensePredictions.clinpredColor) }}</td>
								<td class="w3-left-align"><em class="w3-small">Threshold &ge; {{ thresholds.clinpred }} for Damaging - meta score</em></td>
							</tr>
						{% endif %}
						<tr><!--Meta SVM-->
							<td class="w3-left-align"><a href="{{ urls.dbnsfp }}" target="_blank" title="Visit dbnsfp">Meta SVM</a>:</td>
							<td class="w3-left-align">{{ span_color(external_data.missensePredictions.metaSVMScore, internal_data.missensePredictions.metaSVMColor) }} ({{ external_data.missensePredictions.metaSVMLRRel }})</td>
							<td class="w3-left-align">{{ span_color(external_data.missensePredictions.metaSVMPred, internal_data.missensePredictions.metaSVMColor) }}</td>
							<td class="w3-left-align"><em class="w3-small">Threshold &ge; {{ thresholds['meta-svm'] }} for Damaging (reliabilty index: 0-10), 10:high - meta score</em></td>
						</tr>
						<tr><!--Meta LR-->
							<td class="w3-left-align"><a href="{{ urls.dbnsfp }}" target="_blank" title="Visit dbnsfp">Meta LR</a>:</td>
							<td class="w3-left-align">{{ span_color(external_data.missensePredictions.metaLRScore, internal_data.missensePredictions.metaLRColor) }} ({{ external_data.missensePredictions.metaSVMLRRel }})</td>
							<td class="w3-left-align">{{ span_color(external_data.missensePredictions.metaLRPred, internal_data.missensePredictions.metaLRColor) }}</td>
							<td class="w3-left-align"><em class="w3-small">Threshold &ge; {{ thresholds['meta-lr'] }}  for Damaging (reliabilty index: 0-10), 10:high - meta score</em></td>
						</tr>
						<tr><!--Mistic-->
							<td class="w3-left-align"><a href="{{ urls.mistic }}" target="_blank" title="Visit Mistic">Mistic</a>:</td>
							<td class="w3-left-align">{{ span_color(external_data.missensePredictions.misticScore, internal_data.missensePredictions.misticColor) }}</td>
							<td class="w3-left-align">{{ span_color(external_data.missensePredictions.misticPred, internal_data.missensePredictions.misticColor) }}</td>
							<td class="w3-left-align"><em class="w3-small">Threshold &ge; {{ thresholds['mistic'] }}  for Damaging - meta score</em></td>
						</tr>
					</tbody>
				</table>
				{% if internal_data.missensePredictions.siftStar == '*' or internal_data.missensePredictions.polyphen2HdivStar == '*' or internal_data.missensePredictions.polyphen2HvarStar == '*' or external_data.missensePredictions.fathmmstar == '*' %}
					<p>*No score found for the current isoform. The reported score is the most deleterious found in other isoforms.</p>
				{% endif %}
				<br/><br/>
				<div class="w3-row-padding">
					<!--Missense radar chart-->
					<div class="w3-col m4">
						<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="$('#radar_missense_modal').show();" title="Display a radar chart of missense scores" style="width:100%">Show radar chart</span>
					</div>
				</div>
				<div id="radar_missense_modal" class="w3-modal w3-animate-opacity" style="display:block;z-index:2002;">
					<div class="w3-modal-content w3-card-4 w3-display-middle w3-animate-opacity w3-mobile">
						<header class="w3-container w3-blue">
							<span class="w3-button w3-display-topright w3-xlarge" onclick="$('#radar_missense_modal').hide();">X</span>
							<br /><span class="w3-xlarge">Radar view of missense predictors</span><br />
							<p>Values are normalised (0-1), 0 being the less damaging and 1 the most for each predictor.<span id="total_missense_scores"></span></p>
						</header>
						<div class="w3-container">
							<div class="chart-container" style="position: relative;">
								<canvas id="missense_radar"><p>Radar Chart for missense predictions</p></canvas>
							</div>
						</div>
					</div>
				</div>
				{# set missense predictors values to min when no values to plot radar chart #}
				{% set missense_pred = namespace(sift=external_data.missensePredictions.siftScore, pphvar=external_data.missensePredictions.polyphen2HvarScore, pphdiv=external_data.missensePredictions.polyphen2HdivScore, revel=external_data.missensePredictions.revelScore, clinpred=external_data.missensePredictions.clinpredScore, fathmm=external_data.missensePredictions.fathmmScore, metasvm=external_data.missensePredictions.metaSVMScore, metalr=external_data.missensePredictions.metaLRScore, mistic=external_data.missensePredictions.misticScore, num_aca_single=4, num_non_aca_single=2, num_aca_meta=5, num_non_aca_meta=3) %}
				{% if external_data.missensePredictions.siftPred == 'no prediction' %}
					{% set missense_pred.sift = 1 %}
					{% set missense_pred.num_aca_single = missense_pred.num_aca_single - 1 %}
					{% set missense_pred.num_non_aca_single = missense_pred.num_non_aca_single - 1 %}
				{% endif %}
				{% if external_data.missensePredictions.polyphen2HvarPred == 'no prediction' %}
					{% set missense_pred.pphvar = 0 %}
					{% set missense_pred.num_aca_single = missense_pred.num_aca_single - 1 %}
				{% endif %}
				{% if external_data.missensePredictions.polyphen2HdivPred == 'no prediction' %}
					{% set missense_pred.pphdiv = 0 %}
					{% set missense_pred.num_aca_single = missense_pred.num_aca_single - 1 %}
				{% endif %}
				{% if external_data.missensePredictions.revelPred == 'no prediction' %}
					{% set missense_pred.revel = 0 %}
					{% set missense_pred.num_aca_meta = missense_pred.num_aca_meta - 1 %}
				{% endif %}
				{% if external_data.missensePredictions.clinpredPred == 'no prediction' %}
					{% set missense_pred.clinpred = 0 %}
					{% set missense_pred.num_aca_meta = missense_pred.num_aca_meta - 1 %}
				{% endif %}
				{% if external_data.missensePredictions.fathmmPred == 'no prediction' %}
					{% set missense_pred.fathmm = 10.64 %}
					{% set missense_pred.num_aca_single = missense_pred.num_aca_single - 1 %}
					{% set missense_pred.num_non_aca_single = missense_pred.num_non_aca_single - 1 %}
				{% endif %}
				{% if external_data.missensePredictions.metaSVMPred == 'no prediction' %}
					{% set missense_pred.metasvm = -2 %}
					{% set missense_pred.num_aca_meta = missense_pred.num_aca_meta - 1 %}
					{% set missense_pred.num_non_aca_meta = missense_pred.num_non_aca_meta - 1 %}
				{% endif %}
				{% if external_data.missensePredictions.metaLRPred == 'no prediction' %}
					{% set missense_pred.metalr = 0 %}
					{% set missense_pred.num_aca_meta = missense_pred.num_aca_meta - 1 %}
					{% set missense_pred.num_non_aca_meta = missense_pred.num_non_aca_meta - 1 %}
				{% endif %}
				{% if external_data.missensePredictions.misticPred == 'no prediction' %}
					{% set missense_pred.mistic = 0 %}
					{% set missense_pred.num_aca_meta = missense_pred.num_aca_meta - 1 %}
					{% set missense_pred.num_non_aca_meta = missense_pred.num_non_aca_meta - 1 %}
				{% endif %}
				<script>
					//var ctx = document.getElementById("missense_radar").getContext('2d')
					Chart.defaults.global.defaultFontSize=16;
					var point_font_size = 16
					var wi = $(window).width();
					if (wi < 400) {
						Chart.defaults.global.defaultFontSize=12;
						point_font_size = 12;
					}
					// normalisation function for scores not [0;1]
					function normalizeData(val, my_array) {
						var new_val = val == 0 ? 0 : ((val - my_array[0]) / (my_array[1] - my_array[0])).toFixed(2);
						return parseFloat(new_val);
					}
					var normalisedFathmm = 1-normalizeData({{ missense_pred.fathmm }}, [-16.13,10.64]);
					var normalisedMetasvm = normalizeData({{ missense_pred.metasvm }}, [-2,3]);
					// var numberOfScores = {% if g.user['academic'] or not g.user %} {{ missense_pred.num_aca }} {% else %} {{ missense_pred.num_non_aca }} {% endif %};
					var numberOfScoresSingle = {% if g.user['academic'] or not g.user %} {{ missense_pred.num_aca_single }} {% else %} {{ missense_pred.num_non_aca_single }} {% endif %};
					var numberOfScoresMeta = {% if g.user['academic'] or not g.user %} {{ missense_pred.num_aca_meta }} {% else %} {{ missense_pred.num_non_aca_meta }} {% endif %};
					var totalNormalisedScoresSingle = ((1-{{ missense_pred.sift }}) + {% if g.user['academic'] or not g.user %} ({{ missense_pred.pphdiv }}) + ({{ missense_pred.pphvar }}) + {% endif %} (normalisedFathmm)).toFixed(3);
					var totalNormalisedScoresMeta = ((normalisedMetasvm) + {% if g.user['academic'] or not g.user %} ({{ missense_pred.revel }}) + ({{ missense_pred.clinpred }}) + {% endif %} ({{ missense_pred.metalr }}) + ({{ missense_pred.mistic }})).toFixed(3)
					// var totalNormalisedScores = ((1-{{ missense_pred.sift }}) + {% if g.user['academic'] or not g.user %} ({{ missense_pred.pphdiv }}) + ({{ missense_pred.pphvar }}) + ({{ missense_pred.revel }}) + ({{ missense_pred.clinpred }}) + {% endif %} (normalisedFathmm) + (normalisedMetasvm) + ({{ missense_pred.metalr }}) + ({{ missense_pred.mistic }})).toFixed(3);
					//var meanNormalisedScores = (totalNormalisedScores / numberOfScores).toFixed(3);
					var meanNormalisedScoresSingle = (totalNormalisedScoresSingle / numberOfScoresSingle).toFixed(3);
					var meanNormalisedScoresMeta = (totalNormalisedScoresMeta / numberOfScoresMeta).toFixed(3);
					$('#total_missense_scores').html('<br/>Mean normalised score of all single predictors: ' + meanNormalisedScoresSingle + ' (' + totalNormalisedScoresSingle + ' / ' + numberOfScoresSingle + ') - for meta predictors: ' + meanNormalisedScoresMeta + ' (' + totalNormalisedScoresMeta + ' / ' + numberOfScoresMeta + ')');
					var myRadarChart = new Chart($("#missense_radar"), {
						type: 'radar',
						data: {
							labels: ['SIFT', {% if g.user['academic'] or not g.user %}'Polyphen 2 HDIV', 'Polyphen 2 HVAR', 'REVEL', 'ClinPred', {% endif %}'FATHMM', 'MetaSVM', 'MetaLR', 'Mistic'],
							datasets: [{
								label:'Missense predictors',
								// data: [1-{{ missense_pred.sift }}, {% if g.user['academic'] or not g.user %} {{ missense_pred.pphdiv }}, {{ missense_pred.pphvar }}, {{ missense_pred.revel }}, {{ missense_pred.clinpred }}, {% endif %}1-normalizeData({{ missense_pred.fathmm }}, [-16.13,10.64]), normalizeData({{ missense_pred.metasvm }}, [-2,3]), {{ missense_pred.metalr }}, {{ missense_pred.mistic }}],
								data: [1-{{ missense_pred.sift }}, {% if g.user['academic'] or not g.user %} {{ missense_pred.pphdiv }}, {{ missense_pred.pphvar }}, {{ missense_pred.revel }}, {{ missense_pred.clinpred }}, {% endif %}normalisedFathmm, normalisedMetasvm, {{ missense_pred.metalr }}, {{ missense_pred.mistic }}],
								fill: true,
								backgroundColor:'rgb(255, 99, 132, 0.2)',
								borderColor:'rgb(255, 99, 132)',
								pointBackgroundColor:'rgb(255, 99, 132)',
								pointBorderColor:'#fff',
								pointHoverBackgroundColor:'#fff',
								pointHoverBorderColor:'rgb(255, 99, 132)'
							}]
						},
						options:{elements:{line:{tension:0,borderWidth:3}},scale:{pointLabels:{fontSize:point_font_size},ticks:{beginAtZero:true,min:0,max:1,stepSize:0.2}},multiTooltipTemplate : "<%=label%> : <%=value%>",onAnimationComplete: "done"}
					});
				</script>
				{% if g.user %}
					{{ modal_js('pubmed', 'radar_splicing_modal', 'radar_missense_modal', 'defgen_modal_hg19', 'defgen_modal_hg38', 'modify_class_modal') }}
				{% else %}
					{{ modal_js('pubmed', 'radar_splicing_modal', 'radar_missense_modal', 'defgen_modal_hg19', 'defgen_modal_hg38') }}
				{% endif %}
			</div>


		{% elif internal_data.noMatch.dbnsfp is not none %}
			{{ info(internal_data.noMatch.dbnsfp) }}
			{% if g.user %}
				{% if external_data.VCF.hg19.ncbiChr is not none %}
					{{ modal_js('pubmed', 'radar_splicing_modal', 'defgen_modal_hg19', 'defgen_modal_hg38', 'modify_class_modal') }}
				{% else %}
					{{ modal_js('pubmed', 'radar_splicing_modal', 'defgen_modal_hg38', 'modify_class_modal') }}
				{% endif %}
			{% else %}
				{% if external_data.VCF.hg19.ncbiChr is not none %}
					{{ modal_js('pubmed', 'radar_splicing_modal', 'defgen_modal_hg19', 'defgen_modal_hg38') }}
				{% else %}
					{{ modal_js('pubmed', 'radar_splicing_modal', 'defgen_modal_hg38') }}
				{% endif %}
			{% endif %}
		{% elif external_data.splicingPredictions.spliceai_DS_AG is not none  %}
			{% if g.user %}
				{% if external_data.VCF.hg19.ncbiChr is not none %}
					{{ modal_js('pubmed', 'radar_splicing_modal', 'defgen_modal_hg19', 'defgen_modal_hg38', 'modify_class_modal') }}
				{% else %}
					{{ modal_js('pubmed', 'radar_splicing_modal', 'defgen_modal_hg38', 'modify_class_modal') }}
				{% endif %}
			{% else %}
				{% if external_data.VCF.hg19.ncbiChr is not none %}
					{{ modal_js('pubmed', 'radar_splicing_modal', 'defgen_modal_hg19', 'defgen_modal_hg38') }}
				{% else %}
					{{ modal_js('pubmed', 'radar_splicing_modal', 'defgen_modal_hg38') }}
				{% endif %}
			{% endif %}
		{% else %}
			{% if g.user %}
				{% if external_data.VCF.hg19.ncbiChr is not none %}
					{{ modal_js('pubmed', 'defgen_modal_hg19', 'defgen_modal_hg38', 'modify_class_modal') }}
				{% else %}
					{{ modal_js('pubmed', 'defgen_modal_hg38', 'modify_class_modal') }}
				{% endif %}
			{% else %}
				{% if external_data.VCF.hg19.ncbiChr is not none %}
					{{ modal_js('pubmed', 'defgen_modal_hg19', 'defgen_modal_hg38') }}
				{% else %}
					{{ modal_js('pubmed', 'defgen_modal_hg38') }}
				{% endif %}
			{% endif %}
		{% endif %}
		<!--EpiSign section-->
		{% if internal_data.episignature.ExpertACMGClassification is not none %}
		<hr />
		{{ section_title('episignature', 'Epigenetic Signature') }}
		{{ info("The next table presents DNA methylation 'episignature'. The majority of genomic DNA methylation data has utilized genomic microarrays (currently the Illumina Infinium Methylation arrays). The DNA methylation pattern can be used as biomarker for the diagnosis of genetic diseases using easily accessible tissues such as peripheral blood and may help to classify variants (<a href='%s32644126' target='_blank'>Sadikovic et al. 2020</a>)."|format(urls.ncbi_pubmed)) }}
		<div class="w3-responsive w3-container">
			<table class="w3-table w3-centered display compact" style="width:90%" id="episignature_table">
				<thead>
					<tr>
						<th class="w3-left-align">Features</th>
						<th class="w3-left-align" width="40%">Values</th>
						<th class="w3-left-align">Descriptions</th>
					</tr>
				</thead>
				<tbody>
					<!--WDSTS Episignature-->
					<tr>
						<td class="w3-left-align">DNA Methylation pattern</td>
						<td class="w3-left-align">{{ internal_data.episignature.DNAMethylationPattern }}</td>
						<td class="w3-left-align">{{ internal_data.episignature.Description }}</td>
					</tr>
					<!--ACMG Classification-->
					<tr>
						<td class="w3-left-align">Expert opinion</td>
						<td class="w3-left-align"><span style="color:{{ internal_data.episignature.htmlCode }};">Class {{ internal_data.episignature.ExpertACMGClassification }} ({{ internal_data.episignature.ACMGTranslation }})</span></td>
						<td class="w3-left-align">Expert opinion after DNA methylation study (Anne-Sophie Lebre)</td>
					</tr>
					<!--Reference Article-->
					<tr>
						<td class="w3-left-align">Reference Article</td>
						<td class="w3-left-align">
							{# {% if internal_data.episignature.pubmedID is number and internal_data.episignature.pubmedID > 0 %} #}
							{% if internal_data.episignature.pubmedID %}
									<a href="{{ urls.ncbi_pubmed }}{{ internal_data.episignature.pubmedID }}" target="_blank">{{ internal_data.episignature.referenceArticle }}</a>
							{% else %}
								{{ internal_data.episignature.referenceArticle }}
							{% endif %}
						</td>
						<td class="w3-left-align">reference data</td>
					</tr>
				</tbody>
			</table>
		</div>
		{% endif %}
		<!--Classification section-->
		<hr />
		{{ section_title('class', 'Classification History') }}
		<div class="w3-responsive w3-container">
			{% if g.user %}
			<div class="w3-margin w3-panel w3-sand w3-leftbar w3-display-container">
				<p>By assigning an ACMG class to this variant, you also contribute to the international knowledge on DNA variants, as we will submit the variant and the classification to the <a href="https://databases.lovd.nl/shared/genes/{{ external_data.gene.symbol }}" target="_blank">Global Variome Shared LOVD</a> (Read <a href="{{ url_for('md.about') }}#data_sharing" target="_blank">more</a>).</p>
				<p>
					{% if g.user['lovd_export'] == true %}
					Your LOVD export status is currently <span class="w3-text-green">On</span> - This means the data will be exported.
					{% else %}
					Your LOVD export status is currently <span class="w3-text-red">Off</span> - This means the data will not be exported.
					{% endif %}
					<br /> You can modify it at your <a href="{{ url_for('auth.profile', mobiuser_id=0) }}" target="_blank">profile</a> page.
				</p>
			</div>
			{% endif %}
			<!--ACMG Classifications-->
			<table class="w3-table w3-centered display compact" style="width:90%" id="class_table">
				<thead>
					<tr>
						<th class="w3-left-align">Variant</th>
						<th class="w3-left-align">User</th>
						<th class="w3-left-align">Date</th>
						<th class="w3-left-align">ACMG Classification</th>
						<th class="w3-left-align">Comments</th>
					</tr>
				</thead>
				<tbody>
				{% if class_history is none %}
				<tr id="no_class">
					<td></td>
					<td></td>
					<td></td>
					<td>No classification defined yet for this variant</td>
					<td></td>
				</tr>
				{% else %}
					{% for classif in class_history %}
					<tr id="{{ classif.id }}-{{ classif.acmg_class }}-{{ classif.variant_feature_id }}"><!--classification infos-->
						{% autoescape false %}
						<td>{{ "%s(<em>%s</em>):c.%s" | format(classif.refseq, classif.gene_symbol, classif.c_name) }}
						{% if classif.refseq == external_data.gene.RefSeqTranscript %}
						<div class="w3-tag w3-round w3-green" style="padding:3px">
							<div class="w3-tag w3-round w3-green w3-border w3-border-white">
								Current
							</div>
						</div>
						{% endif %}
						</td>
						{% endautoescape %}
						{% if g.user %}
							{% if g.user['id'] == classif.id %}
							<td class="w3-left-align"><a href="{{ url_for('auth.profile', mobiuser_id=0) }}" target="_blank" title="Visit your profile">{{ classif.username }}</a></td>
							{% elif classif.username != 'mobidetails' %}
							<td class="w3-left-align">
								<a href="{{ url_for('auth.profile', mobiuser_id=classif.id) }}" target="_blank" title="See {{ classif.username }} profile">{{ classif.username }}</a>
								{% if classif.email_pref == true %}
									&nbsp;&nbsp;<i class="w3-xlarge fa fa-envelope-o" onclick="$('#receiver_id').val('{{ classif.id }}');$('#receiver_username').html('{{ classif.username }}');$('#message_modal').show();" title="Send {{ classif.username }} a message via mobidetails" style="cursor:pointer;"></i>
								{% endif %}
							</td>
							{% else %}
							<td class="w3-left-align">{{ classif.username }}</td>
							{% endif %}
						{% else %}
						<td class="w3-left-align">{{ classif.username }}</td>
						{% endif %}
						<td class="w3-left-align">{{ classif.class_date }}</td>
						<td class="w3-left-align"><span style="color:{{ classif.html_code }};">Class {{ classif.acmg_class }} ({{ classif.acmg_translation }})</span></td>
						<td>
							<div class="w3-cell-row">
								<span class="w3-container w3-left-align w3-cell" id="comment_{{ classif.id }}-{{ classif.acmg_class }}-{{ external_data.variantId }}">
								</span>
								{% autoescape false %}
								<script type="application/x-javascript">
										$("#comment_{{ classif.id }}-{{ classif.acmg_class }}-{{ external_data.variantId }}").html(unescape('{{ classif.comment }}'));
								</script>
								{% endautoescape %}
								{% if classif.username == g.user['username'] %}
									{% if run_mode  and run_mode == 'on' %}
									<span class="w3-container w3-right-align w3-cell">
										<i class="fa fa-times w3-xlarge" style="color:red;cursor:pointer;" aria-hidden="true" title="click to remove this classification" onclick="remove_class('{{ external_data.variantId }}', '{{ g.user['id'] }}', '{{ classif.acmg_class }}', '{{ url_for("ajax.remove_class") }}', '{{ csrf_token() }}')"></i>
									</span>
									{% endif %}
								{% endif %}
							</div>
						</td>
					</tr>
					{% endfor %}
				{% endif %}
				</tbody>
			</table><br/><br/>
			{% if g.user %}
				<div class="w3-container" id="message_return"></div>
				<div class="w3-row-padding">
					{% if run_mode  and run_mode == 'on' %}
						<div class="w3-col m4" id="modify_class">
							<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" title="add your own classification to the history" style="width:100%" id="modify_class_btn" onclick="$('#modify_class_modal').show();">Add your classification</span>
							<div class="w3-modal" id="modify_class_modal" style="display:none;">
								<div class="w3-modal-content w3-display-middle w3-animate-opacity w3-mobile">
									<header class="w3-container w3-blue">
										<span class="w3-button w3-display-topright w3-xlarge" onclick="$('#modify_class_modal').hide();">X</span>
										<br />
										<p><span class="w3-xlarge">Add your ACMG classification to the history:</span></p>
										<br />
									</header>
									<div class="w3-container w3-center">
										<form method="post" class="w3-container w3-text-blue w3-margin w3-large" style="width:100%" id="acmg_form" enctype="application/x-www-form-urlencoded">
											 <!--action="{{ url_for('ajax.modif_class') }}"-->
											<!-- <input type="hidden" name="csrf_token" value="{{ csrf_token() }}"/> -->
											<!-- <input type="hidden" name="variant_id" value="{{ external_data.variantId }}" />
											variant id set at the beginning of the page -->
											<!-- <input type="hidden" name="mobiuser_id" value="{{ g.user['id'] }}" /> -->
											<h3>This will be visible to all users:</h3><br />
											<div class="w3-row w3-section">
												<div class="w3-quarter w3-left-align">
													<label for="acmg_select" form="acmg_form">ACMG class:</label>
												</div>
												<div class="w3-rest">
													<select class="w3-select w3-border" name="acmg_select" id="acmg_select" form="acmg_form">
														<option value="1" style="color:#00A020;">Benign</option>
														<option value="2" style="color:#0000A0;">Likely Benign</option>
														<option value="3" style="color:#969696;">Uncertain Significance</option>
														<option value="4" style="color:#FF6020;">Likely Pathogenic</option>
														<option value="5" style="color:#FF0000;">Pathogenic</option>
													</select>
												</div><br /><br />
											</div>
											<div class="w3-row w3-section">
												<div class="w3-quarter w3-left-align">
													<label for="acmg_comment" form="acmg_form">Comments (optional, use english, please, limit 500 chars):</label>
												</div>
												<div class="w3-rest">
													<textarea name="acmg_comment" id="acmg_comment" rows="4" cols="50" maxlength="500" form="acmg_form"></textarea>
												</div>
											</div>
											<span class="w3-center">
												<input type="button" id = "sub" class="w3-button w3-section w3-blue w3-ripple"  onclick="modify_class('{{ external_data.variantId }}', '{{ g.user['id'] }}', '{{ url_for("ajax.modif_class") }}', '{{ csrf_token() }}');" value="submit"/>
												<!--<input type="submit" id = "sub" class="w3-button w3-section w3-blue w3-ripple"  onclick="$('#html').css('cursor', 'progress');$('#acmg_form').css('cursor', 'progress');" value="submit"/>-->
											</span>
										</form>
									</div>
									<br /><br />
								</div>
							</div>
						</div>
					{% else %}
						<span>Please wait for the end of the maintenance to classify this variant.</span>
					{% endif %}
				</div>
				<div class="w3-modal" id="message_modal" style="display:none;">
					<div class="w3-modal-content w3-display-middle w3-animate-opacity w3-mobile">
						<header class="w3-container w3-blue">
							<span class="w3-button w3-display-topright w3-xlarge" onclick="$('#message_modal').hide();">X</span>
							<br />
							<p><span class="w3-xlarge">Use MobiDetails to contact a peer (<span id="receiver_username"></span>):</span></p>
							<br />
						</header>
						<div class="w3-container w3-center">
							<form method="post" class="w3-container w3-text-blue w3-margin w3-large" style="width:100%" id="message_form" enctype="application/x-www-form-urlencoded">
								<input type="hidden" name="csrf_token" value="{{ csrf_token() }}"/>
								<input type="hidden" name="sender_id" id="sender_id" value="{{ g.user['id'] }}" />
								<input type="hidden" name="receiver_id" id="receiver_id" value="1" />
								<span>This message will be sent via MobiDetails. The contact will also receive your email address and be able to answer directly.</span><br />
								<p><span id="message_object">Object: Query via MobiDetails from {{ g.user['username'] }} concerning {{ external_data.gene.RefSeqTranscript }}:c.{{ internal_data.nomenclatures.cName }}</span></p>
								<div class="w3-row w3-section w3-responsive">
									<div class="w3-quarter w3-left-align">
										<label for="message_body" form="message_form"><span >Type your message here:</span></label>
									</div>
									<div class="w3-rest">
										<textarea name="message_body" id="message_body" rows="7" cols="50" maxlength="1000" form="message_form"></textarea>
									</div>
								</div>
								<span class="w3-center">
									<input type="button" id="sub_message" class="w3-button w3-section w3-blue w3-ripple"  onclick="send_var_message('{{ url_for("ajax.send_var_message") }}', '{{ csrf_token() }}');" value="submit"/>
								</span>
							</form>
						</div>
						<br /><br />
					</div>
				</div>
			{% else %}
				{{ info("To modify the classification history, you must be <a href='%s'>logged in</a> (you might want to <a href='%s' target='_blank'>create an account</a> before)."|format(url_for('auth.login'), url_for('auth.register'))) }}
			{% endif %}
		</div>
		<br/><br/>
		<!--eighth section-->
		<hr />
		{{ section_title('admin', 'Administrative information') }}
		<div class="w3-responsive w3-container">
			<table class="w3-table w3-centered display compact" style="width:90%" id="admin_table">
				<thead>
					<tr>
						<th class="w3-left-align">Features</th>
						<th class="w3-left-align">Values</th>
						<th class="w3-left-align">Descriptions</th>
					</tr>
				</thead>
				<tbody>
					<tr><!--creation user-->
						<td class="w3-left-align">Creation user:</td>
						{% if external_data.admin.creationUserName != 'mobidetails' and g.user %}
							<td class="w3-left-align"><a href="{{ url_for('auth.profile', mobiuser_id=internal_data.admin.creationUser) }}" target="_blank" title="Click to see {{ external_data.admin.creationUserName }} profile"><span id="owner_username">{{ external_data.admin.creationUserName }}</span></a></td>
						{% else %}
							<td class="w3-left-align"><span id="owner_username">{{ external_data.admin.creationUserName }}</span></td>
						{% endif %}
						<td class="w3-left-align"><em class="w3-small">User who created the variant</em></td>
					</tr>
					<tr><!--creation date-->
						<td class="w3-left-align">Creation date:</td>
						<td class="w3-left-align">{{ external_data.admin.creationDate }}</td>
						<td class="w3-left-align"><em class="w3-small">Date of creation in MD</em></td>
					</tr>
				</tbody>
			</table><br /><br />
			{% if g.user %}
				<span id="current_user" style="display:none;">{{ g.user['username'] }}</span>
			{% else %}
				<span id="current_user" style="display:none;">mobidetails</span>
			{% endif %}
			{% if external_data.admin.creationUserName != 'mobidetails' and not g.user %}
				{{ info("To contact the creation user, you must be <a href='%s'>logged in</a> (you might want to <a href='%s' target='_blank'>create an account</a> before)."|format(url_for('auth.login'), url_for('auth.register'))) }}
			{% elif external_data.admin.creationUserName == 'mobidetails' %}
				{{ info("The first <a href='%s'>logged in</a> user to assign a classification to this variant will become its \"owner\" (<a href='%s' target='_blank'>create an account</a>)."|format(url_for('auth.login'), url_for('auth.register'))) }}
			{% endif %}
		</div>
		<br/><br/><br />
		<div class="w3-container w3-padding-64 w3-center" id="res_versions" style="display:none">
			<div class="w3-responsive w3-container">
				<table class="w3-table w3-centered display compact" style="width:90%" id="resource_table">
					<caption>References and versions of external resources used in MD:</caption>
					<thead>
						<tr>
							<th class="w3-left-align">Resource</th>
							<th class="w3-left-align">Version</th>
							<th class="w3-left-align">Reference</th>
						</tr>
					</thead>
					<tbody>
						{% for tool in external_tools %}
							<tr>
								<td class="w3-left-align">
									<a href="{{ urls[external_tools[tool]['short_name']] }}" target="_blank" rel="noopener noreferrer" title="Visit {{ tool }}">{{ tool }}</a>
								</td>
								<td class="w3-left-align">{{ external_tools[tool]['version'] }}</td>
								{% if external_tools[tool]['paper'] != 'None' %}
									<td class="w3-left-align"><a href="{{ external_tools[tool]['paper'] }}" target="_blank" rel="noopener noreferrer" title="Paper">{{ external_tools[tool]['firstauthor'] }} et al., {{ external_tools[tool]['year'] }}</a></td>
								{% else %}
									<td class="w3-left-align">No paper available yet</td>
								{% endif %}
							</tr>
						{% endfor %}
					</tbody>
				</table>
			</div>
		</div>
	</div>
	<script>
		$(document).ready(function() {
			litvar2('{{ url_for('ajax.litvar2') }}', '{{ csrf_token() }}');
			{% if external_data.VCF.hg19.ncbiChr is not none %}
				lovd('{{ url_for('ajax.lovd') }}', '{{ csrf_token() }}', $('#genome_19').text(), $('#chrom_19').text(), $('#hg19_g_name').text());
				intervar('{{ url_for('ajax.intervar') }}', '{{ csrf_token() }}');
			{% else %}
				lovd('{{ url_for('ajax.lovd') }}', '{{ csrf_token() }}', $('#genome_38').text(), $('#chrom_38').text(), $('#hg38_g_name').text());
			{% endif %}
			run_spip('{{ url_for('ajax.spip') }}', '{{ external_data.variantId }}', '{{ csrf_token() }}');
			spliceaivisual('{{ url_for('ajax.spliceaivisual') }}', '{{ url_for('static', filename='') }}', 'automatic', '{{ csrf_token() }}')
		});
	</script>
{% endblock %}
